// igniteui-react-charts/esm5/lib/DomainChart_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends, __values } from "tslib";
import { Base, INotifyPropertyChanged_$type, BaseError, Type, PropertyChangedEventArgs, Delegate_$type, fromEnum, String_$type, runOn, delegateCombine, EnumUtil, enumGetBox, typeCast, delegateRemove, markType, TypeRegistrar, getInstanceType } from "igniteui-react-core";
import { IDataLegendTarget_$type } from "igniteui-react-core";
import { IToolbarTarget_$type } from "igniteui-react-core";
import { ArrayExtension } from "igniteui-react-core";
import { List$1 } from "igniteui-react-core";
import { Brush } from "igniteui-react-core";
import { XamDataChart } from "./XamDataChart";
import { DataSeriesType_$type } from "igniteui-react-core";
import { MarkerType_$type } from "./MarkerType";
import { Color } from "igniteui-react-core";
import { ColorUtil } from "igniteui-react-core";
import { DVContainer } from "igniteui-react-core";
import { DataToolTipLayerGroupingMode_$type } from "igniteui-react-core";
import { DataTooltipGroupedPositionX_$type } from "igniteui-react-core";
import { DataTooltipGroupedPositionY_$type } from "igniteui-react-core";
import { DataAbbreviationMode_$type } from "igniteui-react-core";
import { DataLegendValueMode_$type } from "igniteui-react-core";
import { Thickness } from "igniteui-react-core";
import { DataLegendHeaderDateMode_$type } from "igniteui-react-core";
import { DataLegendHeaderTimeMode_$type } from "igniteui-react-core";
import { DataLegendSummaryType_$type } from "igniteui-react-core";
import { LegendItemBadgeShape_$type } from "igniteui-react-core";
import { DataLegendUnitsMode_$type } from "igniteui-react-core";
import { DataLegendLabelMode_$type } from "igniteui-react-core";
import { ChartSortDescriptionCollection } from "./ChartSortDescriptionCollection";
import { FilterExpressionCollection } from "igniteui-react-core";
import { ChartSummaryDescriptionCollection } from "./ChartSummaryDescriptionCollection";
import { SeriesSelectionMode_$type } from "./SeriesSelectionMode";
import { SeriesSelectionBehavior_$type } from "./SeriesSelectionBehavior";
import { ChartSelectedItemCollection } from "./ChartSelectedItemCollection";
import { SortGroupParser } from "igniteui-react-core";
import { FilterExpressionParsingHelper } from "igniteui-react-core";
import { FilterStringErrorsParsingEventArgs } from "./FilterStringErrorsParsingEventArgs";
import { SummaryParser } from "igniteui-react-core";
import { BrushCollection } from "igniteui-react-core";
import { SeriesHighlightedValuesDisplayMode_$type } from "./SeriesHighlightedValuesDisplayMode";
import { Rect } from "igniteui-react-core";
import { HorizontalAlignment_$type } from "igniteui-react-core";
import { UnknownValuePlotting_$type } from "igniteui-react-core";
import { SeriesOutlineMode_$type } from "./SeriesOutlineMode";
import { MarkerOutlineMode_$type } from "./MarkerOutlineMode";
import { MarkerFillMode_$type } from "./MarkerFillMode";
import { MarkerAutomaticBehavior_$type } from "./MarkerAutomaticBehavior";
import { TrendLineType_$type } from "igniteui-react-core";
import { ComputedPlotAreaMarginMode_$type } from "./ComputedPlotAreaMarginMode";
import { SeriesPlotAreaMarginHorizontalMode_$type } from "./SeriesPlotAreaMarginHorizontalMode";
import { SeriesPlotAreaMarginVerticalMode_$type } from "./SeriesPlotAreaMarginVerticalMode";
import { SeriesHighlightingMode_$type } from "./SeriesHighlightingMode";
import { SeriesHighlightingBehavior_$type } from "./SeriesHighlightingBehavior";
import { LegendHighlightingMode_$type } from "./LegendHighlightingMode";
import { LegendItemBadgeMode_$type } from "igniteui-react-core";
import { ToolTipType_$type } from "./ToolTipType";
import { CrosshairsDisplayMode_$type } from "./CrosshairsDisplayMode";
import { SeriesViewerScrollbarMode_$type } from "./SeriesViewerScrollbarMode";
import { SeriesViewerHorizontalScrollbarPosition_$type } from "./SeriesViewerHorizontalScrollbarPosition";
import { SeriesViewerVerticalScrollbarPosition_$type } from "./SeriesViewerVerticalScrollbarPosition";
import { Defaults } from "./Defaults";
import { Series } from "./Series";
import { ICommandStateChangedListener_$type } from "igniteui-react-core";
import { DeviceUtils } from "igniteui-react-core";
import { Dictionary$2 } from "igniteui-react-core";
import { DataSeriesAdapter } from "igniteui-react-core";
import { ChartSortDescription } from "./ChartSortDescription";
import { ChartSummaryDescription } from "./ChartSummaryDescription";
import { ObservableCollection$1 } from "igniteui-react-core";
import { Random } from "igniteui-react-core";
import { ValueLayerValueMode_$type } from "./ValueLayerValueMode";
import { ObservableColorCollection } from "igniteui-react-core";
import { BrushCollectionUtil } from "igniteui-react-core";
import { ChartSeriesEventArgs } from "./ChartSeriesEventArgs";
import { DomainChartSeriesPointerEventArgs } from "./DomainChartSeriesPointerEventArgs";
import { BrushUtil } from "igniteui-react-core";
import { IUnknownValuePlotting_$type } from "igniteui-react-core";
import { ITrendLineSeries_$type } from "./ITrendLineSeries";
import { isNaN_ } from "igniteui-react-core";
import { stringIsNullOrEmpty, stringEmpty, stringStartsWith } from "igniteui-react-core";
/**
 * @hidden
 */
var DomainChart = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(DomainChart, _super);
    function DomainChart() {
        var _this = _super.call(this) || this;
        _this.kz = null;
        _this.kx = null;
        _this.hc = null;
        _this.h4 = NaN;
        _this.kw = null;
        _this.ii = 0;
        _this.ij = 0;
        _this.ik = 0;
        _this.ih = 0;
        _this.ic = 0;
        _this.ie = 0;
        _this.id = 0;
        _this.ib = 0;
        _this.vu = null;
        _this.vv = null;
        _this.h1 = NaN;
        _this.il = NaN;
        _this.ia = NaN;
        _this.ho = NaN;
        _this.kk = 0;
        _this.em = null;
        _this.kb = null;
        _this.g1 = false;
        _this.ce = null;
        _this.lv = null;
        _this.lx = null;
        _this.ea = null;
        _this.fr = false;
        _this.fq = false;
        _this.hg = null;
        _this.propertyChanged = null;
        _this.kj = 0;
        _this.gu = false;
        _this.b7 = null;
        _this.hb = null;
        _this.cb = null;
        _this.ft = false;
        _this.cc = null;
        _this._commandCompleted = null;
        _this._invalidateActions = null;
        _this.g9 = new List$1(ICommandStateChangedListener_$type, 0);
        _this.by = 0;
        _this.hs = NaN;
        _this.ht = NaN;
        _this.hq = DeviceUtils.g(10);
        _this.hr = DeviceUtils.g(10);
        _this.bu = 0;
        _this.bw = 0;
        _this.fk = true;
        _this.e = null;
        _this.b = null;
        _this.d = null;
        _this.a = null;
        _this.be = 0;
        _this.kf = -1;
        _this.kg = -1;
        _this.bs = 0;
        _this.ll = stringEmpty();
        _this.fl = true;
        _this.lm = null;
        _this.f = null;
        _this.w7 = new Thickness(0, DeviceUtils.g(5));
        _this.fm = true;
        _this.ln = "N/A";
        _this.fn = false;
        _this.w8 = new Thickness(1, DeviceUtils.g(2), DeviceUtils.g(0), DeviceUtils.g(2), DeviceUtils.g(0));
        _this.vn = null;
        _this.ex = null;
        _this.lf = null;
        _this.c = null;
        _this.le = stringEmpty();
        _this.bi = 0;
        _this.bk = 0;
        _this.lg = null;
        _this.vf = null;
        _this.w1 = new Thickness(1, DeviceUtils.g(2), DeviceUtils.g(0), DeviceUtils.g(0), DeviceUtils.g(0));
        _this.w0 = new Thickness(0, DeviceUtils.g(5));
        _this.fj = true;
        _this.ep = null;
        _this.ve = null;
        _this.wz = new Thickness(1, DeviceUtils.g(2), DeviceUtils.g(0), DeviceUtils.g(0), DeviceUtils.g(0));
        _this.wy = new Thickness(0, DeviceUtils.g(5));
        _this.fi = false;
        _this.eo = null;
        _this.vi = null;
        _this.es = null;
        _this.bo = 0;
        _this.li = null;
        _this.w4 = new Thickness(1, DeviceUtils.g(2), DeviceUtils.g(0), DeviceUtils.g(0), DeviceUtils.g(0));
        _this.w3 = new Thickness(0, DeviceUtils.g(5));
        _this.vk = null;
        _this.eu = null;
        _this.lh = null;
        _this.vh = null;
        _this.er = null;
        _this.lj = null;
        _this.vj = null;
        _this.et = null;
        _this.wx = new Thickness(0, DeviceUtils.g(0));
        _this.ck = 0;
        _this.bq = 0;
        _this.lk = null;
        _this.vm = null;
        _this.w6 = new Thickness(0, DeviceUtils.g(0));
        _this.ew = null;
        _this.w5 = new Thickness(0, DeviceUtils.g(0));
        _this.vl = null;
        _this.ev = null;
        _this.bm = 0;
        _this.vg = null;
        _this.w2 = new Thickness(1, DeviceUtils.g(8), DeviceUtils.g(0), DeviceUtils.g(0), DeviceUtils.g(0));
        _this.eq = null;
        _this.seriesAdded = null;
        _this.seriesRemoved = null;
        _this.seriesPointerEnter = null;
        _this.seriesPointerLeave = null;
        _this.seriesPointerMove = null;
        _this.seriesPointerDown = null;
        _this.seriesPointerUp = null;
        _this.calloutStyleUpdating = null;
        _this.calloutRenderStyleUpdating = null;
        _this.selectedSeriesItemsChanged = null;
        _this.focusedSeriesItemsChanged = null;
        _this.hf = null;
        _this.e9 = null;
        _this.fc = false;
        _this.g8 = new Dictionary$2(String_$type, Base.$, 0);
        _this.ha = new List$1(String_$type, 0);
        _this.ay = null;
        _this.au = null;
        _this.at = null;
        _this.ec = null;
        _this.ed = null;
        _this.a6 = null;
        _this.ek = null;
        _this.el = null;
        _this.f2 = true;
        _this.f0 = true;
        _this.f3 = true;
        _this.f1 = true;
        _this.fz = true;
        _this.ax = null;
        _this.aw = null;
        _this.av = null;
        _this.ef = null;
        _this.ee = null;
        _this.a5 = null;
        _this.dp = 0;
        _this.dn = 0;
        _this.vt = null;
        _this.vq = null;
        _this.dl = 0;
        _this.ao = new ChartSelectedItemCollection();
        _this.an = new ChartSelectedItemCollection();
        _this.ls = null;
        _this.lo = null;
        _this.lq = null;
        _this.lp = null;
        _this.lr = null;
        _this.filterStringErrorsParsing = null;
        _this.lt = null;
        _this.kv = null;
        _this.ku = null;
        _this.h = null;
        _this.g = null;
        _this.t = null;
        _this.w = null;
        _this.r = null;
        _this.s = null;
        _this.b9 = null;
        _this.fs = true;
        _this.fv = true;
        _this.fx = false;
        _this.f4 = false;
        _this.c9 = 1;
        _this.fu = false;
        _this.wu = new Rect(0, 0, 0, 1, 1);
        _this.lw = null;
        _this.lu = null;
        _this.u5 = 1;
        _this.u4 = 1;
        _this.d6 = 0;
        _this.h9 = NaN;
        _this.ig = NaN;
        _this.df = 0;
        _this.cs = 0;
        _this.cq = 0;
        _this.h2 = NaN;
        _this.h3 = NaN;
        _this.cy = null;
        _this.co = 4;
        _this.u = null;
        _this.v = null;
        _this.ki = 400;
        _this.hn = NaN;
        _this.fb = false;
        _this.x = null;
        _this.d4 = 0;
        _this.im = DeviceUtils.g(1.5);
        _this.fa = true;
        _this.h6 = NaN;
        _this.h8 = NaN;
        _this.h7 = NaN;
        _this.h5 = NaN;
        _this.a9 = 0;
        _this.dh = 0;
        _this.dj = 0;
        _this.dd = 0;
        _this.db = 0;
        _this.cg = 0;
        _this.cl = 0;
        _this.ci = 0;
        _this.kl = 7;
        _this.d2 = 4;
        _this.bc = 0;
        _this.fh = false;
        _this.vd = null;
        _this.hp = NaN;
        _this.vc = null;
        _this.fg = true;
        _this.u8 = null;
        _this.va = null;
        _this.u9 = null;
        _this.vb = null;
        _this.kd = -1;
        _this.ke = -1;
        _this.fw = false;
        _this.fy = true;
        _this.fo = false;
        _this.vo = null;
        _this.vp = null;
        _this.kh = -1;
        _this.fd = false;
        _this.ff = false;
        _this.fe = false;
        _this.hk = null;
        _this.aj = null;
        _this.lc = null;
        _this.ld = null;
        _this.lb = "CalloutLabel";
        _this.la = "CalloutContent";
        _this.d8 = null;
        _this.y = null;
        _this.io = NaN;
        _this.du = 0;
        _this.dv = 0;
        _this.ds = 0;
        _this.d0 = 0;
        _this.vr = null;
        _this.vs = null;
        _this.hy = NaN;
        _this.hx = NaN;
        _this.hu = NaN;
        _this.hv = NaN;
        _this.hw = NaN;
        _this.h0 = NaN;
        _this.hz = NaN;
        _this.fp = true;
        _this.vw = null;
        _this.vx = null;
        _this.is = NaN;
        _this.ir = NaN;
        _this.ip = NaN;
        _this.iv = NaN;
        _this.iq = NaN;
        _this.iu = NaN;
        _this.it = NaN;
        _this.f5 = true;
        _this.ix = NaN;
        _this.iw = NaN;
        _this.iz = NaN;
        _this.iy = NaN;
        _this.hf = new List$1(Delegate_$type, 0);
        _this.tg();
        _this.sw();
        return _this;
    }
    DomainChart.prototype.provideContainer = function (a) {
        if (a == null) {
            this.tx(function (b) {
                b.destroy();
                b.provideContainer(null);
            });
            return;
        }
        var b = this.kz;
        this.kz = a;
        if (this.e9 != null && this.kz != null && b == null) {
            this.kz.setTimeout(this.e9, 0);
            this.e9 = null;
        }
        var c = ((function () {
            var $ret = new XamDataChart();
            $ret.e3 = true;
            return $ret;
        })());
        c.provideContainer(a);
        this.dataChart = c;
        this.th();
    };
    DomainChart.prototype.destroy = function () {
        this.fq = true;
        this.provideContainer(null);
    };
    DomainChart.prototype.th = function () {
        if (this.dataChart == null) {
            return;
        }
        var a = this.dataChart.c8.j.e();
        this.ab = a.a;
        this.af = a.f;
        this.ad = a.b;
        this.ae = a.c;
        this.ah = a.g;
        this.i2 = Defaults.j;
        this.ti();
    };
    DomainChart.prototype.ti = function () {
    };
    DomainChart.prototype.te = function (a) {
        if (a == null) {
            return;
        }
        a.h6 = this.jw;
        a.h4 = this.ju;
        a.h5 = this.jv;
        a.h3 = this.jt;
        a.u3 = this.wq;
        a.h0 = this.jp;
        a.h2 = this.jr;
        a.h1 = this.jq;
        a.hz = this.jo;
        a.u2 = this.wp;
        if (!isNaN_(this.jh)) {
            a.hu = this.jh;
        }
    };
    DomainChart.prototype.tb = function () {
    };
    DomainChart.prototype.tk = function (a, b) {
        a.jt = this.ks;
        a.c3 = this.en;
    };
    DomainChart.prototype.t8 = function (a, b, c) {
        var _this = this;
        switch (a) {
            case "TitleTopMargin":
                if (this.dataChart != null) {
                    this.dataChart.h6 = this.jw;
                }
                break;
            case "TitleLeftMargin":
                if (this.dataChart != null) {
                    this.dataChart.h4 = this.ju;
                }
                break;
            case "PixelScalingRatio":
                if (this.dataChart != null) {
                    this.tx(function (d) { return d.hu = _this.jh; });
                }
                break;
            case "TitleRightMargin":
                if (this.dataChart != null) {
                    this.dataChart.h5 = this.jv;
                }
                break;
            case "TitleBottomMargin":
                if (this.dataChart != null) {
                    this.dataChart.h3 = this.jt;
                }
                break;
            case "TitleTextColor":
                if (this.dataChart != null) {
                    this.dataChart.u3 = this.wq;
                }
                break;
            case "SubtitleLeftMargin":
                if (this.dataChart != null) {
                    this.dataChart.h0 = this.jp;
                }
                break;
            case "SubtitleTopMargin":
                if (this.dataChart != null) {
                    this.dataChart.h2 = this.jr;
                }
                break;
            case "SubtitleRightMargin":
                if (this.dataChart != null) {
                    this.dataChart.h1 = this.jq;
                }
                break;
            case "SubtitleBottomMargin":
                if (this.dataChart != null) {
                    this.dataChart.hz = this.jo;
                }
                break;
            case "SubtitleTextColor":
                if (this.dataChart != null) {
                    this.dataChart.u2 = this.wp;
                }
                break;
            case "LeftMargin":
                if (this.dataChart != null) {
                    this.dataChart.ht = this.je;
                }
                break;
            case "TopMargin":
                if (this.dataChart != null) {
                    this.dataChart.h7 = this.jx;
                }
                break;
            case "RightMargin":
                if (this.dataChart != null) {
                    this.dataChart.hy = this.jn;
                }
                break;
            case "BottomMargin":
                if (this.dataChart != null) {
                    this.dataChart.hg = this.i1;
                }
                break;
            case "TransitionDuration":
                this.s7(function (d) { return d.jt = _this.ks; });
                break;
            case "TransitionEasingFunction":
                this.s7(function (d) { return d.c3 = _this.en; });
                break;
        }
    };
    Object.defineProperty(DomainChart.prototype, "tooltipTemplate", {
        get: function () {
            return ArrayExtension.getModulus$1(Base.$, this.tooltipTemplates, 0);
        },
        set: function (a) {
            this.tooltipTemplates = a == null ? null : new List$1(Base.$, 1, [a]);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "tooltipTemplates", {
        get: function () {
            return this.hc;
        },
        set: function (a) {
            var b = this.tooltipTemplates;
            if (a != b) {
                this.hc = a;
                this.t5("TooltipTemplates", b, this.tooltipTemplates);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "jh", {
        get: function () {
            return this.h4;
        },
        set: function (a) {
            var b = this.h4;
            if (a != b) {
                this.h4 = a;
                this.t5("PixelScalingRatio", b, this.jh);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "toolTip", {
        get: function () {
            return this.kw;
        },
        set: function (a) {
            var b = this.toolTip;
            if (b != a) {
                this.kw = a;
                this.t5("ToolTip", b, this.toolTip);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "ju", {
        get: function () {
            return this.ii;
        },
        set: function (a) {
            var b = this.ju;
            if (a != b) {
                this.ii = a;
                this.t5("TitleLeftMargin", b, this.ju);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "jv", {
        get: function () {
            return this.ij;
        },
        set: function (a) {
            var b = this.jv;
            if (a != b) {
                this.ij = a;
                this.t5("TitleRightMargin", b, this.jv);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "jw", {
        get: function () {
            return this.ik;
        },
        set: function (a) {
            var b = this.jw;
            if (a != b) {
                this.ik = a;
                this.t5("TitleTopMargin", b, this.jw);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "jt", {
        get: function () {
            return this.ih;
        },
        set: function (a) {
            var b = this.jt;
            if (a != b) {
                this.ih = a;
                this.t5("TitleBottomMargin", b, this.jt);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "jp", {
        get: function () {
            return this.ic;
        },
        set: function (a) {
            var b = this.jp;
            if (a != b) {
                this.ic = a;
                this.t5("SubtitleLeftMargin", b, this.jp);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "jr", {
        get: function () {
            return this.ie;
        },
        set: function (a) {
            var b = this.jr;
            if (a != b) {
                this.ie = a;
                this.t5("SubtitleTopMargin", b, this.jr);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "jq", {
        get: function () {
            return this.id;
        },
        set: function (a) {
            var b = this.jq;
            if (a != b) {
                this.id = a;
                this.t5("SubtitleRightMargin", b, this.jq);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "jo", {
        get: function () {
            return this.ib;
        },
        set: function (a) {
            var b = this.jo;
            if (a != b) {
                this.ib = a;
                this.t5("SubtitleBottomMargin", b, this.jo);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "wp", {
        get: function () {
            return this.vu;
        },
        set: function (a) {
            var b = this.wp;
            if (a != b) {
                this.vu = a;
                this.t5("SubtitleTextColor", b, this.wp);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "wq", {
        get: function () {
            return this.vv;
        },
        set: function (a) {
            var b = this.wq;
            if (a != b) {
                this.vv = a;
                this.t5("TitleTextColor", b, this.wq);
            }
        },
        enumerable: false,
        configurable: true
    });
    DomainChart.prototype.get_je = function () {
        return this.h1;
    };
    DomainChart.prototype.set_je = function (a) {
        var b = this.je;
        if (a != b) {
            this.h1 = a;
            this.t5("LeftMargin", b, this.je);
        }
    };
    Object.defineProperty(DomainChart.prototype, "je", {
        get: function () {
            return this.get_je();
        },
        set: function (a) {
            this.set_je(a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "jx", {
        get: function () {
            return this.il;
        },
        set: function (a) {
            var b = this.jx;
            if (a != b) {
                this.il = a;
                this.t5("TopMargin", b, this.jx);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "jn", {
        get: function () {
            return this.ia;
        },
        set: function (a) {
            var b = this.jn;
            if (a != b) {
                this.ia = a;
                this.t5("RightMargin", b, this.jn);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "i1", {
        get: function () {
            return this.ho;
        },
        set: function (a) {
            var b = this.i1;
            if (a != b) {
                this.ho = a;
                this.t5("BottomMargin", b, this.i1);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "ks", {
        get: function () {
            return this.kk;
        },
        set: function (a) {
            var b = this.ks;
            if (a != b) {
                this.kk = a;
                this.t5("TransitionDuration", b, this.ks);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "en", {
        get: function () {
            return this.em;
        },
        set: function (a) {
            var b = this.en;
            if (a != b) {
                this.em = a;
                this.t5("TransitionEasingFunction", b, this.en);
            }
        },
        enumerable: false,
        configurable: true
    });
    DomainChart.prototype.tf = function (a) {
        if (this.dataChart == null) {
            return;
        }
        a.lv = this.rq;
        a.lj = this.rf;
    };
    DomainChart.prototype.tl = function (a, b) {
        this.tm(a, b);
    };
    DomainChart.prototype.tc = function () {
    };
    DomainChart.prototype.tm = function (a, b) {
        if (a == null) {
            return;
        }
        if (this.d3 == 3) {
            return;
        }
        var toolTipTemplate_ = ArrayExtension.getModulus$1(Base.$, this.tooltipTemplates, b);
        if (toolTipTemplate_ == null) {
            a.f9 = true;
        }
        else {
            var series_ = a;
            a.f9 = false;
            series_.externalObject.tooltipTemplate = toolTipTemplate_;
        }
    };
    DomainChart.prototype.u0 = function (a, b, c) {
    };
    DomainChart.prototype.k1 = function (a) {
        return a;
        return null;
    };
    DomainChart.prototype.kc = function (a) {
        return null;
    };
    DomainChart.prototype.t9 = function (a, b, c) {
        var _this = this;
        switch (a) {
            case "SubtitleTextStyle":
                if (this.dataChart != null) {
                    this.dataChart.lj = this.rf;
                }
                break;
            case "TitleTextStyle":
                if (this.dataChart != null) {
                    this.dataChart.lv = this.rq;
                }
                break;
            case "TooltipTemplate":
                if (this.g1) {
                    return;
                }
                this.g1 = true;
                this.tooltipTemplate = this.k1(this.tooltipTemplate);
                this.g1 = false;
                if (this.dataChart != null) {
                    this.dataChart.i1 = this.tooltipTemplate;
                }
                this.toolTip = this.tooltipTemplate;
                break;
            case "ToolTip":
                this.s8(function (d, e) { return _this.tm(d, e); });
                break;
            case "TooltipTemplates":
                if (this.tooltipTemplates == null) {
                    break;
                }
                this.g1 = true;
                this.s8(function (d, e) {
                    var f = _this.k1(ArrayExtension.getModulus$1(Base.$, _this.tooltipTemplates, e));
                    var series_ = d;
                    var templ_ = f;
                    series_.externalObject.tooltipTemplate = templ_;
                });
                this.g1 = false;
                break;
        }
    };
    Object.defineProperty(DomainChart.prototype, "cf", {
        get: function () {
            return this.ce;
        },
        set: function (a) {
            var b = this.cf;
            if (a != b) {
                this.ce = a;
                this.t5("Widget", b, this.cf);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "rf", {
        get: function () {
            return this.lv;
        },
        set: function (a) {
            var b = this.rf;
            if (a != b) {
                this.lv = a;
                this.t5("SubtitleTextStyle", b, this.rf);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "rq", {
        get: function () {
            return this.lx;
        },
        set: function (a) {
            var b = this.rq;
            if (a != b) {
                this.lx = a;
                this.t5("TitleTextStyle", b, this.rq);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "dataChart", {
        get: function () {
            return this.ea;
        },
        set: function (a) {
            var b = this.dataChart;
            if (b != a) {
                this.ea = a;
                this.t5("DataChart", b, this.dataChart);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "isDetached", {
        get: function () {
            return this.fr || this.fq;
        },
        enumerable: false,
        configurable: true
    });
    DomainChart.prototype.onDetachedFromUI = function () {
        this.fr = true;
        this.tx(function (a) { return a.onDetachedFromUI(); });
    };
    DomainChart.prototype.onAttachedToUI = function () {
        this.fr = false;
        this.tx(function (a) { return a.onAttachedToUI(); });
    };
    DomainChart.prototype.tg = function () {
        var e_1, _a, e_2, _b;
        this.hg = new List$1(MarkerType_$type, 0);
        this.bh = new DataSeriesAdapter().e;
        this.ay = new ChartSortDescriptionCollection();
        this.au = new ChartSortDescriptionCollection();
        this.at = new ChartSortDescriptionCollection();
        this.a6 = new ChartSummaryDescriptionCollection();
        this.ec = new FilterExpressionCollection();
        this.ed = new FilterExpressionCollection();
        this.a4.e = this.bh.sortDescriptions;
        this.az.e = this.bh.groupDescriptions;
        this.eg.syncTarget = this.bh.filterExpressions;
        this.eh.syncTarget = this.bh.filterExpressions;
        this.a8.e = this.bh.summaryDescriptions;
        this.a0.e = this.bh.postSortDescriptions;
        this.ek = new SortGroupParser();
        this.el = new SummaryParser();
        this.ax = ((function () {
            var $ret = new ChartSortDescriptionCollection();
            $ret.i = true;
            return $ret;
        })());
        this.aw = ((function () {
            var $ret = new ChartSortDescriptionCollection();
            $ret.i = true;
            return $ret;
        })());
        this.av = ((function () {
            var $ret = new ChartSortDescriptionCollection();
            $ret.i = true;
            return $ret;
        })());
        this.a5 = ((function () {
            var $ret = new ChartSummaryDescriptionCollection();
            $ret.i = true;
            return $ret;
        })());
        this.ee = ((function () {
            var $ret = new FilterExpressionCollection();
            $ret.shouldDetachOnTargetChange = true;
            return $ret;
        })());
        this.ef = ((function () {
            var $ret = new FilterExpressionCollection();
            $ret.shouldDetachOnTargetChange = true;
            return $ret;
        })());
        if (this.fz) {
            if (this.eg.k.count == 0) {
                this.ee.syncTarget = this.bh.filterExpressions;
            }
            this.fz = false;
        }
        if (this.f1) {
            if (this.eh.k.count == 0) {
                this.ef.syncTarget = this.bh.highlightFilterExpressions;
            }
            this.f1 = false;
        }
        if (this.f2) {
            if (this.a4.j.count == 0) {
                this.ax.e = this.bh.sortDescriptions;
            }
            if (this.a0.j.count == 0) {
                this.aw.e = this.bh.postSortDescriptions;
            }
            this.f2 = false;
        }
        if (this.f0) {
            if (this.az.j.count == 0) {
                this.av.e = this.bh.groupDescriptions;
            }
            this.f0 = false;
        }
        if (this.f3) {
            if (this.a8.j.count == 0) {
                this.a5.e = this.bh.summaryDescriptions;
            }
            this.f3 = false;
        }
        if (this.q != null) {
            this.bh.includedProperties = this.q;
        }
        if (this.o != null) {
            this.bh.excludedProperties = this.o;
        }
        this.bh.supportedSeriesTypes = this.hi.toArray();
        this.bh.rules.clear();
        try {
            for (var _c = __values(fromEnum(this.hh)), _d = _c.next(); !_d.done; _d = _c.next()) {
                var a = _d.value;
                this.bh.rules.add(a);
            }
        }
        catch (e_1_1) {
            e_1 = { error: e_1_1 };
        }
        finally {
            try {
                if (_d && !_d.done && (_a = _c.return))
                    _a.call(_c);
            }
            finally {
                if (e_1)
                    throw e_1.error;
            }
        }
        var b = new List$1(String_$type, 0);
        try {
            for (var _e = __values(fromEnum(this.bh.rules)), _f = _e.next(); !_f.done; _f = _e.next()) {
                var c = _f.value;
                b.add(getInstanceType(c).typeName);
            }
        }
        catch (e_2_1) {
            e_2 = { error: e_2_1 };
        }
        finally {
            try {
                if (_f && !_f.done && (_b = _e.return))
                    _b.call(_e);
            }
            finally {
                if (e_2)
                    throw e_2.error;
            }
        }
        var d = this.bh.dataSeries;
        d.collectionChanged = delegateCombine(d.collectionChanged, runOn(this, this.t0));
    };
    DomainChart.prototype.gw = function (a) {
        return a != 24 && this.hi.contains(a);
    };
    DomainChart.prototype.notifyResized = function () {
        this.tx(function (a) { return a.notifyContainerResized(); });
    };
    DomainChart.prototype.tv = function () {
        this.tx(function (a) {
            for (var b = 0; b < a.series.count; b++) {
                a.series._inner[b].pj();
            }
        });
    };
    DomainChart.prototype.s1 = function () {
        if (this.dataChart == null) {
            return;
        }
        if (this.gu) {
            this.ut();
        }
        this.dataChart.flush();
    };
    DomainChart.prototype.s9 = function () {
        if (this.dataChart == null) {
            return;
        }
        this.dataChart.nh();
    };
    DomainChart.prototype.ug = function (a) {
        if (this.propertyChanged != null) {
            this.propertyChanged(this, new PropertyChangedEventArgs(a));
        }
    };
    DomainChart.prototype.s3 = function (a) {
        this.s4(function (b, c) { return a(b); });
    };
    DomainChart.prototype.s4 = function (a) {
        if (this.dataChart == null) {
            return;
        }
        for (var b = 0; b < this.dataChart.series.count; b++) {
            a(this.dataChart.series._inner[b], b);
        }
    };
    DomainChart.prototype.s7 = function (a) {
        this.s8(function (b, c) { return a(b); });
    };
    DomainChart.prototype.s8 = function (a) {
        if (this.dataChart == null) {
            return;
        }
        for (var b = 0; b < this.dataChart.series.count; b++) {
            var c = this.dataChart.series._inner[b];
            if (!c.ec) {
                a(c, b);
            }
        }
    };
    DomainChart.prototype.s5 = function (a) {
        this.s6(function (b, c) { return a(b); });
    };
    DomainChart.prototype.s6 = function (a) {
        if (this.dataChart == null) {
            return;
        }
        for (var b = 0; b < this.dataChart.series.count; b++) {
            var c = this.dataChart.series._inner[b];
            if (c.ed) {
                a(c, b);
            }
        }
    };
    DomainChart.prototype.uq = function () {
        this.s3(function (a) { return a.qq(); });
    };
    DomainChart.prototype.tx = function (a) {
        this.ty(a, true);
    };
    DomainChart.prototype.ty = function (a, b) {
        this.tz(a, b, null);
    };
    DomainChart.prototype.tz = function (a, b, c) {
        if (this.dataChart == null) {
            return;
        }
        a(this.dataChart);
    };
    DomainChart.prototype.gy = function (a) {
        if (!this.gw(a.suggestedSeries)) {
            return false;
        }
        var b = a.data;
        if (b == null) {
            return false;
        }
        return true;
    };
    DomainChart.prototype.gp = function () {
        if (this.dataChart == null) {
            return false;
        }
        return true;
    };
    DomainChart.prototype.pe = function (a) {
        var b = "series_" + (this.dataChart.series.count + 1) + "_" + EnumUtil.getName(DataSeriesType_$type, a.suggestedSeries);
        return b;
    };
    DomainChart.prototype.pf = function (a, b) {
        if (this.gs && this.g6) {
            var c = "";
            var d = 5 + Math.round(DomainChart.k8.nextDouble() * 4);
            for (var e = 0; e < d; e++) {
                c = c + "\u2588";
            }
            return c;
        }
        if (stringIsNullOrEmpty(a.title)) {
            var f = EnumUtil.getName(DataSeriesType_$type, b);
            return f + (this.dataChart.series.count + 1);
        }
        else {
            return a.title;
        }
    };
    DomainChart.prototype.pd = function () {
        return stringEmpty();
    };
    DomainChart.prototype.ut = function () {
        var e_3, _a, e_4, _b;
        if (this.dataChart == null) {
            return;
        }
        this.gu = false;
        this.uv();
        if (this.actualItemsSource == null) {
            return;
        }
        var a = new List$1(MarkerType_$type, 0);
        try {
            for (var _c = __values(fromEnum(this.bh.dataSeries)), _d = _c.next(); !_d.done; _d = _c.next()) {
                var b = _d.value;
                a.add(this.cu(b.suggestedMarker));
            }
        }
        catch (e_3_1) {
            e_3 = { error: e_3_1 };
        }
        finally {
            try {
                if (_d && !_d.done && (_a = _c.return))
                    _a.call(_c);
            }
            finally {
                if (e_3)
                    throw e_3.error;
            }
        }
        this.hg = a;
        var c = 0;
        try {
            for (var _e = __values(fromEnum(this.bh.dataSeries)), _f = _e.next(); !_f.done; _f = _e.next()) {
                var d = _f.value;
                if (!this.gy(d)) {
                    continue;
                }
                if (d.index < 0) {
                    d.index = c;
                }
                this.ur(d);
                this.us(d);
                c++;
            }
        }
        catch (e_4_1) {
            e_4 = { error: e_4_1 };
        }
        finally {
            try {
                if (_f && !_f.done && (_b = _e.return))
                    _b.call(_e);
            }
            finally {
                if (e_4)
                    throw e_4.error;
            }
        }
        this.u1();
        this.s4(runOn(this, this.tj));
    };
    DomainChart.prototype.c8 = function (seriesType_) {
        return (this.externalObject.createSeries(seriesType_).i);
    };
    DomainChart.prototype.c0 = function () {
        return this.c8(28);
    };
    DomainChart.prototype.c2 = function () {
        return this.c8(27);
    };
    DomainChart.prototype.c4 = function () {
        return this.c8(29);
    };
    DomainChart.prototype.c5 = function () {
        return this.c8(25);
    };
    DomainChart.prototype.c1 = function () {
        return this.c8(26);
    };
    DomainChart.prototype.c3 = function () {
        return this.c8(32);
    };
    DomainChart.prototype.c7 = function () {
        return this.c8(33);
    };
    Object.defineProperty(DomainChart.prototype, "b8", {
        get: function () {
            if (TypeRegistrar.isRegistered("AnnotationLayerProxy")) {
                return TypeRegistrar.create("AnnotationLayerProxy");
            }
            else {
                throw new BaseError(1, "annotation layer module is not loaded, but is required.");
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "gl", {
        get: function () {
            return (TypeRegistrar.isRegistered("AnnotationLayerProxy"));
        },
        enumerable: false,
        configurable: true
    });
    DomainChart.prototype.gn = function (a) {
        if (!this.gl) {
            return false;
        }
        return this.b8.matchesType(a);
    };
    DomainChart.prototype.u1 = function () {
        var a = this.he("", true);
        if (a != null) {
            for (var b = 0; b < a.count; b++) {
                this.dataChart.series.add(a._inner[b]);
            }
        }
    };
    DomainChart.prototype.gq = function () {
        return this.bd == 0 || this.bd == 4 || this.bd == 2;
    };
    DomainChart.prototype.gr = function () {
        return this.bd == 0 || this.bd == 4 || this.bd == 3;
    };
    DomainChart.prototype.get_go = function () {
        return this.f9;
    };
    Object.defineProperty(DomainChart.prototype, "go", {
        get: function () {
            return this.get_go();
        },
        enumerable: false,
        configurable: true
    });
    DomainChart.prototype.he = function (a, b) {
        var e_5, _a, e_6, _b;
        if (!this.gl) {
            return null;
        }
        var c = new List$1(Series.$, 0);
        if (b) {
            var d = 1;
            if (this.ff) {
                if (this.hk == null) {
                    if (this.lc != null || this.ld != null || this.lb != null || this.la != null) {
                        try {
                            for (var _c = __values(fromEnum(this.dataChart.series)), _d = _c.next(); !_d.done; _d = _c.next()) {
                                var e = _d.value;
                                var f = this.c0();
                                this.b8.setTargetSeries(f, e);
                                this.b8.setCollisionChannel(f, "MainChartCollisionChannel");
                                this.uz(f, e, this.bh.dataSeries._inner[d - 1]);
                                f.name = a + getInstanceType(f).typeName + d.toString();
                                c.add(f);
                                d++;
                            }
                        }
                        catch (e_5_1) {
                            e_5 = { error: e_5_1 };
                        }
                        finally {
                            try {
                                if (_d && !_d.done && (_a = _c.return))
                                    _a.call(_c);
                            }
                            finally {
                                if (e_5)
                                    throw e_5.error;
                            }
                        }
                        ;
                    }
                }
                else {
                    var g = this.c0();
                    this.uz(g, null, this.bh.dataSeries._inner[d - 1]);
                    g.name = a + getInstanceType(g).typeName + d.toString();
                    if (this.ff && this.fd) {
                        this.b8.setCollisionChannel(g, "MainChartCollisionChannel");
                    }
                    c.add(g);
                }
            }
            if (this.fd) {
                var h = this.c0();
                this.sv(h, this.g5);
                h.name = a + getInstanceType(h).typeName + d.toString();
                if (this.ff && this.fd) {
                    this.b8.setCollisionChannel(h, "MainChartCollisionChannel");
                }
                c.add(h);
            }
        }
        switch (this.bc) {
            case 0:
            case 4:
            case 3:
            case 2:
                {
                    var i = this.c2();
                    this.b8.setCrosshairThickness(i, this.hp);
                    this.b8.setCrosshairUseInterpolation(i, !this.fh);
                    this.b8.setCrosshairHorizontalLineVisibility(i, this.gq() ? 0 : 1);
                    this.b8.setCrosshairHorizontalLineStroke(i, this.vc);
                    this.b8.setCrosshairVerticalLineVisibility(i, this.gr() ? 0 : 1);
                    this.b8.setCrosshairVerticalLineStroke(i, this.vd);
                    this.b8.setCrosshairOnAxesEnabled(i, this.fg);
                    this.b8.setCrosshairXAxisBackground(i, this.u8);
                    this.b8.setCrosshairXAxisForeground(i, this.u9);
                    this.b8.setCrosshairYAxisBackground(i, this.va);
                    this.b8.setCrosshairYAxisForeground(i, this.vb);
                    this.b8.setCrosshairXAxisPrecision(i, this.kd);
                    this.b8.setCrosshairYAxisPrecision(i, this.ke);
                    c.add(i);
                    break;
                }
            case 1:
                {
                    break;
                }
        }
        if (this.fo) {
            var j = this.c4();
            this.b8.setFinalValueBackground(j, this.vo);
            this.b8.setFinalValueForeground(j, this.vp);
            this.b8.setFinalValuePrecision(j, this.kh);
            c.add(j);
        }
        if (b) {
            switch (this.d2) {
                case 1:
                    {
                        var k = this.c5();
                        c.add(k);
                        break;
                    }
                case 2:
                    {
                        var l = this.c1();
                        c.add(l);
                        break;
                    }
                case 4:
                    {
                        var m = this.c3();
                        c.add(m);
                        break;
                    }
                case 3:
                    {
                        this.s3(function (n) { return n.f9 = false; });
                        break;
                    }
                case 0:
                    {
                        this.s3(function (n) { return n.f9 = true; });
                        break;
                    }
            }
        }
        if (this.d9 != null && this.d9.count > 0) {
            for (var n = 0; n < this.d9.count; n++) {
                var o = this.c7();
                this.b8.setValueMode(o, this.d9._inner[n]);
                this.b8.setValueAxisAnnotationEnabled(o, true);
                if (this.ai != null && n < this.ai.count) {
                    this.b8.setBrush(o, this.ai.item(n));
                }
                c.add(o);
            }
        }
        try {
            for (var _e = __values(fromEnum(c)), _f = _e.next(); !_f.done; _f = _e.next()) {
                var p = _f.value;
                if (stringIsNullOrEmpty(p.name) == false) {
                    continue;
                }
                p.name = a + getInstanceType(p).typeName;
            }
        }
        catch (e_6_1) {
            e_6 = { error: e_6_1 };
        }
        finally {
            try {
                if (_f && !_f.done && (_b = _e.return))
                    _b.call(_e);
            }
            finally {
                if (e_6)
                    throw e_6.error;
            }
        }
        return c;
    };
    DomainChart.prototype.sv = function (a, b) {
        this.b8.setIsAutoCalloutBehaviorEnabled(a, true);
        this.b8.setGreedyCollisionMode(a, true);
        this.b8.setUseDedicatedLanes(a, b);
    };
    DomainChart.prototype.uz = function (a, b, c) {
        a.itemsSource = this.hk != null ? this.hk : b.itemsSource;
        this.b8.setXMemberPath(a, this.lc);
        this.b8.setYMemberPath(a, this.ld);
        this.b8.setLabelMemberPath(a, this.lb);
        this.b8.setContentMemberPath(a, this.la);
        if (this.aj != null) {
            this.b8.setCalloutsAllowedPositions(a, this.aj);
        }
    };
    DomainChart.prototype.uv = function () {
        this.uw(this.dataChart);
    };
    DomainChart.prototype.uw = function (a) {
        var e_7, _a;
        var _this = this;
        if (a != null && a == this.dataChart) {
            this.uu();
        }
        if (a == null || a.series.count == 0) {
            return;
        }
        this.s3(function (b) {
            if (_this.gn(b) && _this.b8.isCalloutLayer(b)) {
                _this.b8.removeCalloutStyleUpdating(b, runOn(_this, _this.ue));
            }
            if (_this.gn(b) && _this.b8.isCalloutLayer(b)) {
                _this.b8.removeCalloutRenderStyleUpdating(b, runOn(_this, _this.ud));
            }
            _this.uo(b);
        });
        try {
            for (var _b = __values(fromEnum(a.series)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var b = _c.value;
                this.ux(b);
            }
        }
        catch (e_7_1) {
            e_7 = { error: e_7_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_7)
                    throw e_7.error;
            }
        }
        a.series.clear();
    };
    DomainChart.prototype.ux = function (a) {
        a.itemsSource = null;
        a.legend = null;
        a.n4();
    };
    DomainChart.prototype.t0 = function (a, b) {
        if (this.dataChart == null) {
            return;
        }
        var c = b.oldItems == null ? "null" : b.oldItems.count.toString();
        var d = b.newItems == null ? "null" : b.newItems.count.toString();
        if (this.gu) {
            return;
        }
        var e = this.uc(runOn(this, this.sz), this.gu);
        this.gu = e.p1;
    };
    DomainChart.prototype.sz = function () {
        this.ut();
    };
    DomainChart.prototype.wh = function (a) {
        return ArrayExtension.getModulus$1(Brush.$, this.z, a);
    };
    DomainChart.prototype.wk = function (a) {
        return ArrayExtension.getModulus$1(Brush.$, this.aa, a);
    };
    DomainChart.prototype.wi = function (a) {
        return ArrayExtension.getModulus$1(Brush.$, this.ad, a);
    };
    DomainChart.prototype.wj = function (a) {
        return ArrayExtension.getModulus$1(Brush.$, this.ae, a);
    };
    DomainChart.prototype.wt = function (a) {
        var b = new Color();
        if (this.z == null || this.z.count == 0) {
            b = ArrayExtension.getModulus$1(Color.$, this.hd, a);
        }
        else {
            var c = ArrayExtension.getModulus$1(Brush.$, this.z, a);
            if (c == null) {
                b = ArrayExtension.getModulus$1(Color.$, this.hd, a);
            }
            else {
                b = ColorUtil.q(c);
            }
        }
        return b;
    };
    DomainChart.prototype.e8 = function (a) {
        var b = new ObservableColorCollection();
        b.add(ColorUtil.v(a, 0.5));
        b.add(a);
        return b;
    };
    DomainChart.prototype.ac = function (a) {
        var b = this.e8(a);
        var c = BrushCollectionUtil.a(b);
        return c;
    };
    Object.defineProperty(DomainChart.prototype, "hd", {
        get: function () {
            if (this.hb != null) {
                return this.hb;
            }
            this.hb = new List$1(Color.$, 0);
            this.hb.add(ColorUtil.r("#8BDC5C"));
            this.hb.add(ColorUtil.r("#8961A9"));
            this.hb.add(ColorUtil.r("#6DB1FF"));
            this.hb.add(ColorUtil.r("#82E9D9"));
            this.hb.add(ColorUtil.r("#EA3C63"));
            this.hb.add(ColorUtil.r("#735656"));
            this.hb.add(ColorUtil.r("#F8CE4F"));
            this.hb.add(ColorUtil.r("#A8A8B7"));
            this.hb.add(ColorUtil.r("#E051A9"));
            this.hb.add(ColorUtil.r("#FF903B"));
            return this.hb;
        },
        enumerable: false,
        configurable: true
    });
    DomainChart.prototype.get_g5 = function () {
        return false;
    };
    Object.defineProperty(DomainChart.prototype, "g5", {
        get: function () {
            return this.get_g5();
        },
        enumerable: false,
        configurable: true
    });
    DomainChart.prototype.k0 = function () {
        if (this.dataChart == null) {
            return null;
        }
        return this.dataChart.x4();
    };
    DomainChart.prototype.to = function (a, b) {
    };
    DomainChart.prototype.tp = function (a, b, c) {
    };
    DomainChart.prototype.cu = function (a) {
        if (a == 2) {
            return 3;
        }
        else if (a == 12) {
            return 13;
        }
        else if (a == 6) {
            return 7;
        }
        else if (a == 8) {
            return 9;
        }
        else if (a == 11) {
            return 12;
        }
        else if (a == 1) {
            return 1;
        }
        else if (a == 7) {
            return 8;
        }
        else if (a == 10) {
            return 11;
        }
        else if (a == 4) {
            return 5;
        }
        else if (a == 5) {
            return 6;
        }
        else if (a == 9) {
            return 10;
        }
        else if (a == 3) {
            return 4;
        }
        else if (a == 13) {
            return 0;
        }
        return 2;
    };
    DomainChart.prototype.attachDataLegend = function (a) {
        this.su(a);
    };
    DomainChart.prototype.su = function (a) {
        this.cb = a;
        if (this.dataChart != null) {
            if (!this.ft) {
                this.ft = true;
                this.dataChart.attachDataLegend(a);
            }
        }
    };
    DomainChart.prototype.detachDataLegend = function (a) {
        this.sy(a);
    };
    DomainChart.prototype.sy = function (a) {
        if (this.dataChart != null) {
            this.ft = false;
            this.dataChart.detachDataLegend(a);
        }
        this.cb = null;
    };
    DomainChart.prototype.getDataLegendSeries = function () {
        return this.p();
    };
    DomainChart.prototype.p = function () {
        if (this.dataChart != null) {
            return this.dataChart.getDataLegendSeries();
        }
        else {
            return new Array(0);
        }
    };
    DomainChart.prototype.u2 = function (a) {
        if (this.dataChart != null) {
            this.dataChart.rh(a);
        }
    };
    DomainChart.prototype.u3 = function (a) {
        if (this.dataChart != null) {
            this.dataChart.ri(a);
        }
    };
    DomainChart.prototype.uy = function () {
        if (this.dataChart != null) {
            this.dataChart.qq();
        }
    };
    DomainChart.prototype.get_gx = function () {
        return false;
    };
    Object.defineProperty(DomainChart.prototype, "gx", {
        get: function () {
            return this.get_gx();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "cd", {
        get: function () {
            if (this.cc == null) {
                this.cc = DVContainer.instance.createInstance(IDomainChartToolbarProvider_$type, function () { return "Domain chart toolbar module is not loaded."; });
                this.cc.container = this.kz;
                this.cc.commandChanged = runOn(this, this.t6);
            }
            return this.cc;
        },
        enumerable: false,
        configurable: true
    });
    DomainChart.prototype.t6 = function (a) {
        var e_8, _a;
        try {
            for (var _b = __values(fromEnum(this.g9)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var b = _c.value;
                b.onCommandStateChanged(a.c, a.a, a.b);
            }
        }
        catch (e_8_1) {
            e_8 = { error: e_8_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_8)
                    throw e_8.error;
            }
        }
    };
    DomainChart.prototype.getDesiredToolbarActions = function () {
        if (!this.gx) {
            return null;
        }
        if (this.dataChart != null) {
            var a = this.dataChart.getDesiredToolbarActions();
            return this.cd.getDesiredToolbarActions(this, a);
        }
        return null;
    };
    DomainChart.prototype.onToolCommandExecuting = function (a) {
        if (!this.gx) {
            return 1;
        }
        return this.cd.onToolCommandExecuting(this, a);
    };
    Object.defineProperty(DomainChart.prototype, "commandCompleted", {
        get: function () {
            return this._commandCompleted;
        },
        set: function (a) {
            this._commandCompleted = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "invalidateActions", {
        get: function () {
            return this._invalidateActions;
        },
        set: function (a) {
            this._invalidateActions = a;
        },
        enumerable: false,
        configurable: true
    });
    DomainChart.prototype.tn = function () {
        if (this.invalidateActions != null) {
            this.invalidateActions();
        }
    };
    DomainChart.prototype.addCommandAvailabilityListener = function (a) {
    };
    DomainChart.prototype.removeCommandAvailabilityListener = function (a) {
    };
    DomainChart.prototype.addCommandStateChangedListener = function (a) {
        if (!this.g9.contains(a)) {
            this.g9.add(a);
        }
    };
    DomainChart.prototype.removeCommandStateChangedListener = function (a) {
        if (this.g9.contains(a)) {
            this.g9.remove(a);
        }
    };
    DomainChart.prototype.tu = function (source_, a, b, c) {
        if (this.bh != null) {
            this.bh.notifySetItem(source_, a, b, c);
        }
        if (this.dataChart != null) {
            this.dataChart.n0(source_, a, b, c);
        }
    };
    DomainChart.prototype.tq = function (source_) {
        if (this.bh != null) {
            this.bh.notifyClearItems(source_);
        }
        if (this.dataChart != null) {
            this.dataChart.nr(source_);
        }
    };
    DomainChart.prototype.tr = function (source_, a, b) {
        if (this.bh != null) {
            this.bh.notifyInsertItem(source_, a, b);
        }
        if (this.dataChart != null) {
            this.dataChart.ny(source_, a, b);
        }
    };
    DomainChart.prototype.ts = function (source_, a, b) {
        if (this.bh != null) {
            this.bh.notifyRemoveItem(source_, a, b);
        }
        if (this.dataChart != null) {
            this.dataChart.nz(source_, a, b);
        }
    };
    Object.defineProperty(DomainChart.prototype, "bz", {
        get: function () {
            return this.by;
        },
        set: function (a) {
            var b = this.by;
            this.by = a;
            if (b != this.by) {
                this.t5("DataToolTipGroupingMode", enumGetBox(DataToolTipLayerGroupingMode_$type, b), enumGetBox(DataToolTipLayerGroupingMode_$type, this.by));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "i5", {
        get: function () {
            return this.hs;
        },
        set: function (a) {
            var b = this.hs;
            this.hs = a;
            if (b != this.hs) {
                this.t5("DataToolTipPositionOffsetX", b, this.hs);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "i6", {
        get: function () {
            return this.ht;
        },
        set: function (a) {
            var b = this.ht;
            this.ht = a;
            if (b != this.ht) {
                this.t5("DataToolTipPositionOffsetY", b, this.ht);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "i3", {
        get: function () {
            return this.hq;
        },
        set: function (a) {
            var b = this.hq;
            this.hq = a;
            if (b != this.hq) {
                this.t5("DataToolTipDefaultPositionOffsetX", b, this.hq);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "i4", {
        get: function () {
            return this.hr;
        },
        set: function (a) {
            var b = this.hr;
            this.hr = a;
            if (b != this.hr) {
                this.t5("DataToolTipDefaultPositionOffsetY", b, this.hr);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "bv", {
        get: function () {
            return this.bu;
        },
        set: function (a) {
            var b = this.bu;
            this.bu = a;
            if (b != this.bu) {
                this.t5("DataToolTipGroupedPositionModeX", enumGetBox(DataTooltipGroupedPositionX_$type, b), enumGetBox(DataTooltipGroupedPositionX_$type, this.bu));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "bx", {
        get: function () {
            return this.bw;
        },
        set: function (a) {
            var b = this.bw;
            this.bw = a;
            if (b != this.bw) {
                this.t5("DataToolTipGroupedPositionModeY", enumGetBox(DataTooltipGroupedPositionY_$type, b), enumGetBox(DataTooltipGroupedPositionY_$type, this.bw));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "gg", {
        get: function () {
            return this.fk;
        },
        set: function (a) {
            var b = this.fk;
            this.fk = a;
            if (b != this.fk) {
                this.t5(DomainChart.nx, b, this.fk);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "m", {
        get: function () {
            return this.e;
        },
        set: function (a) {
            var b = this.e;
            this.e = a;
            if (b != this.e) {
                this.t5("DataToolTipIncludedSeries", b, this.e);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "j", {
        get: function () {
            return this.b;
        },
        set: function (a) {
            var b = this.b;
            this.b = a;
            if (b != this.b) {
                this.t5("DataToolTipExcludedSeries", b, this.b);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "l", {
        get: function () {
            return this.d;
        },
        set: function (a) {
            var b = this.d;
            this.d = a;
            if (b != this.d) {
                this.t5("DataToolTipIncludedColumns", b, this.d);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "i", {
        get: function () {
            return this.a;
        },
        set: function (a) {
            var b = this.a;
            this.a = a;
            if (b != this.a) {
                this.t5("DataToolTipExcludedColumns", b, this.a);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "bf", {
        get: function () {
            return this.be;
        },
        set: function (a) {
            var b = this.be;
            this.be = a;
            if (b != this.be) {
                this.t5("DataToolTipValueFormatAbbreviation", enumGetBox(DataAbbreviationMode_$type, b), enumGetBox(DataAbbreviationMode_$type, this.be));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "ko", {
        get: function () {
            return this.kf;
        },
        set: function (a) {
            var b = this.kf;
            this.kf = a;
            if (b != this.kf) {
                this.t5("DataToolTipValueFormatMaxFractions", b, this.kf);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "kp", {
        get: function () {
            return this.kg;
        },
        set: function (a) {
            var b = this.kg;
            this.kg = a;
            if (b != this.kg) {
                this.t5("DataToolTipValueFormatMinFractions", b, this.kg);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "bt", {
        get: function () {
            return this.bs;
        },
        set: function (a) {
            var b = this.bs;
            this.bs = a;
            if (b != this.bs) {
                this.t5("DataToolTipValueFormatMode", enumGetBox(DataLegendValueMode_$type, b), enumGetBox(DataLegendValueMode_$type, this.bs));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "op", {
        get: function () {
            return this.ll;
        },
        set: function (a) {
            var b = this.ll;
            this.ll = a;
            if (b != this.ll) {
                this.t5("DataToolTipValueFormatCulture", b, this.ll);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "gh", {
        get: function () {
            return this.fl;
        },
        set: function (a) {
            var b = this.fl;
            this.fl = a;
            if (b != this.fl) {
                this.t5("DataToolTipValueFormatUseGrouping", b, this.fl);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "ov", {
        get: function () {
            return this.lm;
        },
        set: function (a) {
            var b = this.lm;
            this.lm = a;
            if (b != this.lm) {
                this.t5("DataToolTipValueFormatString", b, this.lm);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "n", {
        get: function () {
            return this.f;
        },
        set: function (a) {
            var b = this.f;
            this.f = a;
            if (b != this.f) {
                this.t5("DataToolTipValueFormatSpecifiers", b, this.f);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "xj", {
        get: function () {
            return this.w7;
        },
        set: function (a) {
            var b = this.w7;
            this.w7 = a;
            if (b != this.w7) {
                this.t5("DataToolTipValueRowMargin", b, this.w7);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "gi", {
        get: function () {
            return this.fm;
        },
        set: function (a) {
            var b = this.fm;
            this.fm = a;
            if (b != this.fm) {
                this.t5("DataToolTipValueRowVisible", b, this.fm);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "o4", {
        get: function () {
            return this.ln;
        },
        set: function (a) {
            var b = this.ln;
            this.ln = a;
            if (b != this.ln) {
                this.t5("DataToolTipValueTextWhenMissingData", b, this.ln);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "gj", {
        get: function () {
            return this.fn;
        },
        set: function (a) {
            var b = this.fn;
            this.fn = a;
            if (b != this.fn) {
                this.t5("DataToolTipValueTextUseSeriesColors", b, this.fn);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "xk", {
        get: function () {
            return this.w8;
        },
        set: function (a) {
            var b = this.w8;
            this.w8 = a;
            if (b != this.w8) {
                this.t5("DataToolTipValueTextMargin", b, this.w8);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "wd", {
        get: function () {
            return this.vn;
        },
        set: function (a) {
            var b = this.vn;
            this.vn = a;
            if (b != this.vn) {
                this.t5("DataToolTipValueTextColor", b, this.vn);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "e7", {
        get: function () {
            return this.ex;
        },
        set: function (a) {
            var b = this.ex;
            this.ex = a;
            if (b != this.ex) {
                this.t5("DataToolTipValueTextStyle", b, this.ex);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "nf", {
        get: function () {
            return this.lf;
        },
        set: function (a) {
            var b = this.lf;
            this.lf = a;
            if (b != this.lf) {
                this.t5("DataToolTipHeaderFormatString", b, this.lf);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "k", {
        get: function () {
            return this.c;
        },
        set: function (a) {
            var b = this.c;
            this.c = a;
            if (b != this.c) {
                this.t5("DataToolTipHeaderFormatSpecifiers", b, this.c);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "nb", {
        get: function () {
            return this.le;
        },
        set: function (a) {
            var b = this.le;
            this.le = a;
            if (b != this.le) {
                this.t5("DataToolTipHeaderFormatCulture", b, this.le);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "bj", {
        get: function () {
            return this.bi;
        },
        set: function (a) {
            var b = this.bi;
            this.bi = a;
            if (b != this.bi) {
                this.t5("DataToolTipHeaderFormatDate", enumGetBox(DataLegendHeaderDateMode_$type, b), enumGetBox(DataLegendHeaderDateMode_$type, this.bi));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "bl", {
        get: function () {
            return this.bk;
        },
        set: function (a) {
            var b = this.bk;
            this.bk = a;
            if (b != this.bk) {
                this.t5("DataToolTipHeaderFormatTime", enumGetBox(DataLegendHeaderTimeMode_$type, b), enumGetBox(DataLegendHeaderTimeMode_$type, this.bk));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "nk", {
        get: function () {
            return this.lg;
        },
        set: function (a) {
            var b = this.lg;
            this.lg = a;
            if (b != this.lg) {
                this.t5("DataToolTipHeaderText", b, this.lg);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "v5", {
        get: function () {
            return this.vf;
        },
        set: function (a) {
            var b = this.vf;
            this.vf = a;
            if (b != this.vf) {
                this.t5("DataToolTipHeaderTextColor", b, this.vf);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "xd", {
        get: function () {
            return this.w1;
        },
        set: function (a) {
            var b = this.w1;
            this.w1 = a;
            if (b != this.w1) {
                this.t5("DataToolTipHeaderTextMargin", b, this.w1);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "xc", {
        get: function () {
            return this.w0;
        },
        set: function (a) {
            var b = this.w0;
            this.w0 = a;
            if (b != this.w0) {
                this.t5("DataToolTipHeaderRowMargin", b, this.w0);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "gf", {
        get: function () {
            return this.fj;
        },
        set: function (a) {
            var b = this.fj;
            this.fj = a;
            if (b != this.fj) {
                this.t5("DataToolTipHeaderRowVisible", b, this.fj);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "ez", {
        get: function () {
            return this.ep;
        },
        set: function (a) {
            var b = this.ep;
            this.ep = a;
            if (b != this.ep) {
                this.t5("DataToolTipHeaderTextStyle", b, this.ep);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "v4", {
        get: function () {
            return this.ve;
        },
        set: function (a) {
            var b = this.ve;
            this.ve = a;
            if (b != this.ve) {
                this.t5("DataToolTipGroupTextColor", b, this.ve);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "xb", {
        get: function () {
            return this.wz;
        },
        set: function (a) {
            var b = this.wz;
            this.wz = a;
            if (b != this.wz) {
                this.t5("DataToolTipGroupTextMargin", b, this.wz);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "xa", {
        get: function () {
            return this.wy;
        },
        set: function (a) {
            var b = this.wy;
            this.wy = a;
            if (b != this.wy) {
                this.t5("DataToolTipGroupRowMargin", b, this.wy);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "ge", {
        get: function () {
            return this.fi;
        },
        set: function (a) {
            var b = this.fi;
            this.fi = a;
            if (b != this.fi) {
                this.t5("DataToolTipGroupRowVisible", b, this.fi);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "ey", {
        get: function () {
            return this.eo;
        },
        set: function (a) {
            var b = this.eo;
            this.eo = a;
            if (b != this.eo) {
                this.t5("DataToolTipGroupTextStyle", b, this.eo);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "v8", {
        get: function () {
            return this.vi;
        },
        set: function (a) {
            var b = this.vi;
            this.vi = a;
            if (b != this.vi) {
                this.t5("DataToolTipSummaryTitleTextColor", b, this.vi);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "e2", {
        get: function () {
            return this.es;
        },
        set: function (a) {
            var b = this.es;
            this.es = a;
            if (b != this.es) {
                this.t5("DataToolTipSummaryTitleTextStyle", b, this.es);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "bp", {
        get: function () {
            return this.bo;
        },
        set: function (a) {
            var b = this.bo;
            this.bo = a;
            if (b != this.bo) {
                this.t5("DataToolTipSummaryType", enumGetBox(DataLegendSummaryType_$type, b), enumGetBox(DataLegendSummaryType_$type, this.bo));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "n3", {
        get: function () {
            return this.li;
        },
        set: function (a) {
            var b = this.li;
            this.li = a;
            if (b != this.li) {
                this.t5("DataToolTipSummaryTitleText", b, this.li);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "xg", {
        get: function () {
            return this.w4;
        },
        set: function (a) {
            var b = this.w4;
            this.w4 = a;
            if (b != this.w4) {
                this.t5("DataToolTipSummaryTitleTextMargin", b, this.w4);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "xf", {
        get: function () {
            return this.w3;
        },
        set: function (a) {
            var b = this.w3;
            this.w3 = a;
            if (b != this.w3) {
                this.t5("DataToolTipSummaryRowMargin", b, this.w3);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "wa", {
        get: function () {
            return this.vk;
        },
        set: function (a) {
            var b = this.vk;
            this.vk = a;
            if (b != this.vk) {
                this.t5("DataToolTipSummaryValueTextColor", b, this.vk);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "e4", {
        get: function () {
            return this.eu;
        },
        set: function (a) {
            var b = this.eu;
            this.eu = a;
            if (b != this.eu) {
                this.t5("DataToolTipSummaryValueTextStyle", b, this.eu);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "ny", {
        get: function () {
            return this.lh;
        },
        set: function (a) {
            var b = this.lh;
            this.lh = a;
            if (b != this.lh) {
                this.t5("DataToolTipSummaryLabelText", b, this.lh);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "v7", {
        get: function () {
            return this.vh;
        },
        set: function (a) {
            var b = this.vh;
            this.vh = a;
            if (b != this.vh) {
                this.t5("DataToolTipSummaryLabelTextColor", b, this.vh);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "e1", {
        get: function () {
            return this.er;
        },
        set: function (a) {
            var b = this.er;
            this.er = a;
            if (b != this.er) {
                this.t5("DataToolTipSummaryLabelTextStyle", b, this.er);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "n9", {
        get: function () {
            return this.lj;
        },
        set: function (a) {
            var b = this.lj;
            this.lj = a;
            if (b != this.lj) {
                this.t5("DataToolTipSummaryUnitsText", b, this.lj);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "v9", {
        get: function () {
            return this.vj;
        },
        set: function (a) {
            var b = this.vj;
            this.vj = a;
            if (b != this.vj) {
                this.t5("DataToolTipSummaryUnitsTextColor", b, this.vj);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "e3", {
        get: function () {
            return this.et;
        },
        set: function (a) {
            var b = this.et;
            this.et = a;
            if (b != this.et) {
                this.t5("DataToolTipSummaryUnitsTextStyle", b, this.et);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "w9", {
        get: function () {
            return this.wx;
        },
        set: function (a) {
            var b = this.wx;
            this.wx = a;
            if (b != this.wx) {
                this.t5("DataToolTipBadgeMargin", b, this.wx);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "cm", {
        get: function () {
            return this.ck;
        },
        set: function (a) {
            var b = this.ck;
            this.ck = a;
            if (b != this.ck) {
                this.t5("DataToolTipBadgeShape", enumGetBox(LegendItemBadgeShape_$type, b), enumGetBox(LegendItemBadgeShape_$type, this.ck));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "br", {
        get: function () {
            return this.bq;
        },
        set: function (a) {
            var b = this.bq;
            this.bq = a;
            if (b != this.bq) {
                this.t5("DataToolTipUnitsDisplayMode", enumGetBox(DataLegendUnitsMode_$type, b), enumGetBox(DataLegendUnitsMode_$type, this.bq));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "oj", {
        get: function () {
            return this.lk;
        },
        set: function (a) {
            var b = this.lk;
            this.lk = a;
            if (b != this.lk) {
                this.t5("DataToolTipUnitsText", b, this.lk);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "wc", {
        get: function () {
            return this.vm;
        },
        set: function (a) {
            var b = this.vm;
            this.vm = a;
            if (b != this.vm) {
                this.t5("DataToolTipUnitsTextColor", b, this.vm);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "xi", {
        get: function () {
            return this.w6;
        },
        set: function (a) {
            var b = this.w6;
            this.w6 = a;
            if (b != this.w6) {
                this.t5("DataToolTipUnitsTextMargin", b, this.w6);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "e6", {
        get: function () {
            return this.ew;
        },
        set: function (a) {
            var b = this.ew;
            this.ew = a;
            if (b != this.ew) {
                this.t5("DataToolTipUnitsTextStyle", b, this.ew);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "xh", {
        get: function () {
            return this.w5;
        },
        set: function (a) {
            var b = this.w5;
            this.w5 = a;
            if (b != this.w5) {
                this.t5("DataToolTipTitleTextMargin", b, this.w5);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "wb", {
        get: function () {
            return this.vl;
        },
        set: function (a) {
            var b = this.vl;
            this.vl = a;
            if (b != this.vl) {
                this.t5("DataToolTipTitleTextColor", b, this.vl);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "e5", {
        get: function () {
            return this.ev;
        },
        set: function (a) {
            var b = this.ev;
            this.ev = a;
            if (b != this.ev) {
                this.t5("DataToolTipTitleTextStyle", b, this.ev);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "bn", {
        get: function () {
            return this.bm;
        },
        set: function (a) {
            var b = this.bm;
            this.bm = a;
            if (b != this.bm) {
                this.t5("DataToolTipLabelDisplayMode", enumGetBox(DataLegendLabelMode_$type, b), enumGetBox(DataLegendLabelMode_$type, this.bm));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "v6", {
        get: function () {
            return this.vg;
        },
        set: function (a) {
            var b = this.vg;
            this.vg = a;
            if (b != this.vg) {
                this.t5("DataToolTipLabelTextColor", b, this.vg);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "xe", {
        get: function () {
            return this.w2;
        },
        set: function (a) {
            var b = this.w2;
            this.w2 = a;
            if (b != this.w2) {
                this.t5("DataToolTipLabelTextMargin", b, this.w2);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "e0", {
        get: function () {
            return this.eq;
        },
        set: function (a) {
            var b = this.eq;
            this.eq = a;
            if (b != this.eq) {
                this.t5("DataToolTipLabelTextStyle", b, this.eq);
            }
        },
        enumerable: false,
        configurable: true
    });
    DomainChart.prototype.ui = function (a) {
        if (this.seriesAdded != null && a != null) {
            this.seriesAdded(this, new ChartSeriesEventArgs(a));
        }
    };
    DomainChart.prototype.uo = function (a) {
        if (this.seriesRemoved != null && a != null) {
            this.seriesRemoved(this, new ChartSeriesEventArgs(a));
        }
    };
    DomainChart.prototype.uk = function (a, b) {
        if (this.seriesPointerEnter != null && a != null) {
            this.seriesPointerEnter(this, new DomainChartSeriesPointerEventArgs(a, b));
        }
    };
    DomainChart.prototype.ul = function (a, b) {
        if (this.seriesPointerLeave != null && a != null) {
            this.seriesPointerLeave(this, new DomainChartSeriesPointerEventArgs(a, b));
        }
    };
    DomainChart.prototype.um = function (a, b) {
        if (this.seriesPointerMove != null && a != null) {
            this.seriesPointerMove(this, new DomainChartSeriesPointerEventArgs(a, b));
        }
    };
    DomainChart.prototype.uj = function (a, b) {
        if (this.seriesPointerDown != null && a != null) {
            this.seriesPointerDown(this, new DomainChartSeriesPointerEventArgs(a, b));
        }
    };
    DomainChart.prototype.un = function (a, b) {
        if (this.seriesPointerUp != null && a != null) {
            this.seriesPointerUp(this, new DomainChartSeriesPointerEventArgs(a, b));
        }
    };
    DomainChart.prototype.ue = function (a, b) {
        if (this.calloutStyleUpdating != null && a != null) {
            this.calloutStyleUpdating(this, b);
        }
    };
    DomainChart.prototype.ud = function (a, b) {
        if (this.calloutRenderStyleUpdating != null && a != null) {
            this.calloutRenderStyleUpdating(this, b);
        }
    };
    DomainChart.prototype.uh = function (a, b) {
        if (this.selectedSeriesItemsChanged != null && a != null) {
            this.selectedSeriesItemsChanged(this, b);
        }
    };
    DomainChart.prototype.uf = function (a, b) {
        if (this.focusedSeriesItemsChanged != null && a != null) {
            this.focusedSeriesItemsChanged(this, b);
        }
    };
    DomainChart.prototype.ub = function (a) {
        var _this = this;
        if (!this.hf.contains(a)) {
            this.hf.add(a);
            var b = false;
            var c = this.uc(function () {
                a();
                _this.hf.remove(a);
            }, b);
            b = c.p1;
        }
    };
    DomainChart.prototype.uc = function (a, b) {
        if (!b) {
            b = true;
            if (this.kz != null) {
                this.kz.setTimeout(a, 0);
            }
            else {
                this.e9 = a;
            }
        }
        return {
            p1: b
        };
    };
    DomainChart.prototype.get_gz = function () {
        return true;
    };
    Object.defineProperty(DomainChart.prototype, "gz", {
        get: function () {
            return this.get_gz();
        },
        enumerable: false,
        configurable: true
    });
    DomainChart.prototype.td = function (a) {
        var e_9, _a, e_10, _b;
        var _this = this;
        var b = a == this.dataChart;
        if (b) {
            a.ag = this.z;
        }
        if (b) {
            a.aj = this.aa;
        }
        a.ah = this.ad;
        a.ai = this.ae;
        a.bg = this.cp;
        a.legend = this.legend;
        a.ws = this.gt;
        a.wv = this.g0 && this.gz;
        a.fy = this.g4;
        if (b) {
            if (this.an != null && this.an.count > 0) {
                try {
                    for (var _c = __values(fromEnum(this.an)), _d = _c.next(); !_d.done; _d = _c.next()) {
                        var c = _d.value;
                        a.focusedSeriesItems.add(c);
                    }
                }
                catch (e_9_1) {
                    e_9 = { error: e_9_1 };
                }
                finally {
                    try {
                        if (_d && !_d.done && (_a = _c.return))
                            _a.call(_c);
                    }
                    finally {
                        if (e_9)
                            throw e_9.error;
                    }
                }
            }
            if (this.ao != null && this.ao.count > 0) {
                try {
                    for (var _e = __values(fromEnum(this.ao)), _f = _e.next(); !_f.done; _f = _e.next()) {
                        var d = _f.value;
                        a.selectedSeriesItems.add(d);
                    }
                }
                catch (e_10_1) {
                    e_10 = { error: e_10_1 };
                }
                finally {
                    try {
                        if (_f && !_f.done && (_b = _e.return))
                            _b.call(_e);
                    }
                    finally {
                        if (e_10)
                            throw e_10.error;
                    }
                }
            }
            a.ch = this.dw;
            a.ci = this.dx;
            a.hl = this.i7;
            a.uv = this.wl;
            a.hm = this.i8;
            a.hn = this.i9;
            a.ho = this.ja;
            a.uw = this.wm;
            a.ce = this.dt;
            a.et = this.gm;
            a.hp = this.jb;
            a.hq = this.jc;
            a.hr = this.jd;
            a.h8 = this.j0;
            a.u5 = this.wr;
            a.ie = this.j6;
            a.h9 = this.j1;
            a.ia = this.j2;
            a.u6 = this.ws;
            a.cl = this.d1;
            a.gb = this.g7;
            a.ib = this.j3;
            a.ic = this.j4;
            a.id = this.j5;
            a.ii = this.j7;
            a.ij = this.j8;
            a.il = this.ka;
            a.ik = this.j9;
            if (a.cb != this.dr || a.b6 != this.dm) {
                a.c8.b4(this.kz);
                a.cb = this.dr;
                a.b6 = this.dm;
            }
            if (a.b9 != this.dq) {
                a.c8.b4(this.kz);
                a.b9 = this.dq;
            }
            if (this.wo != null) {
                a.u1 = this.wo;
            }
            if (this.wg != null) {
                a.uk = this.wg;
            }
            a.selectedSeriesItemsChanged = delegateCombine(a.selectedSeriesItemsChanged, runOn(this, this.uh));
            a.focusedSeriesItemsChanged = delegateCombine(a.focusedSeriesItemsChanged, runOn(this, this.uf));
        }
        a.wd = this.di;
        a.wg = this.dk;
        if (this.cb != null) {
            if (!this.ft) {
                this.su(this.cb);
            }
        }
        if (a.by != this.de || a.bw != this.dc) {
            a.c8.b4(this.kz);
            a.by = this.de;
            a.bw = this.dc;
        }
        a.v6 = this.bb();
        if (!isNaN_(this.jj)) {
            a.xx = this.jj;
        }
        if (!isNaN_(this.jl)) {
            a.xz = this.jl;
        }
        if (!isNaN_(this.jk)) {
            a.xy = this.jk;
        }
        if (!isNaN_(this.ji)) {
            a.xw = this.ji;
        }
        a.bd = this.ch;
        a.seriesMouseEnter = delegateCombine(a.seriesMouseEnter, function (e, f) { return _this.uk(f.series, f.item); });
        a.seriesMouseLeave = delegateCombine(a.seriesMouseLeave, function (e, f) { return _this.ul(f.series, f.item); });
        a.seriesMouseMove = delegateCombine(a.seriesMouseMove, function (e, f) { return _this.um(f.series, f.item); });
        a.seriesMouseLeftButtonDown = delegateCombine(a.seriesMouseLeftButtonDown, function (e, f) { return _this.uj(f.series, f.item); });
        a.seriesMouseLeftButtonUp = delegateCombine(a.seriesMouseLeftButtonUp, function (e, f) { return _this.un(f.series, f.item); });
        if (b) {
            a.vx = this.ww;
        }
        if (b) {
            a.lo = this.rj;
        }
        a.ue = this.u7;
        if (b) {
            a.lc = this.q8;
        }
        a.ud = this.u6;
        a.wo = this.f6;
        a.eg = this.f7;
        a.ht = this.je;
        a.hy = this.jn;
        a.h7 = this.jx;
        a.hg = this.i1;
        if (a == this.dataChart) {
            this.ut();
        }
        this.te(a);
        this.tf(a);
        this.tn();
    };
    DomainChart.prototype.ag = function (a) {
        var e_11, _a;
        if (!this.g3 || !this.g6 || this.itemsSource != null) {
            return a;
        }
        if (a == null) {
            return null;
        }
        var b = new BrushCollection();
        try {
            for (var _b = __values(fromEnum(a)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var c = _c.value;
                var d = BrushUtil.t(c);
                b.add(d);
            }
        }
        catch (e_11_1) {
            e_11 = { error: e_11_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_11)
                    throw e_11.error;
            }
        }
        return b;
    };
    DomainChart.prototype.wn = function (a) {
        if (!this.g3 || !this.g6 || this.itemsSource != null) {
            return a;
        }
        var b = BrushUtil.h(0, 0, 0, 0);
        return b;
    };
    DomainChart.prototype.tj = function (a, b) {
        if (!isNaN_(this.jm)) {
            a.iz = this.jm;
        }
        if (!isNaN_(this.js) && !a.ed) {
            a.i5 = this.js;
        }
        if (!isNaN_(this.i2) && a.ea) {
            a.i5 = this.i2;
        }
        if (!isNaN_(this.jz) && a.ee) {
            a.i5 = this.jz;
        }
        a.legendItemBadgeShape = this.cn;
        a.a5 = this.cj;
        a.b8 = this.dg;
        a.re(this.ct);
        a.rd(this.cr);
        a.iw = this.jf;
        if (!isNaN_(this.jg)) {
            a.rf(this.jg);
        }
        if (!isNaN_(this.i0)) {
            a.h0 = this.i0;
        }
        if (a.isUsableInLegend) {
            a.legend = this.legend;
        }
        if (typeCast(IUnknownValuePlotting_$type, a) !== null) {
            a.unknownValuePlotting = this.d7;
        }
        a.ey = this.gv;
        a.bm = this.da;
        if (this.gn(a) && this.b8.isCalloutLayer(a)) {
            this.b8.setCalloutStyleUpdatingEventEnabled(a, this.ga);
            this.b8.addCalloutStyleUpdating(a, runOn(this, this.ue));
        }
        if (this.gn(a) && this.b8.isCalloutLayer(a)) {
            this.b8.setCalloutRenderStyleUpdatingEventEnabled(a, this.ga);
            this.b8.addCalloutRenderStyleUpdating(a, runOn(this, this.ud));
        }
        if (this.gn(a) && this.b8.isDataToolTipLayer(a)) {
            this.sq(a);
        }
        this.ss(a, b);
        this.st(a, b);
        this.tk(a, b);
        this.tl(a, b);
        this.ui(a);
    };
    Object.defineProperty(DomainChart.prototype, "f8", {
        get: function () {
            return this.fc;
        },
        enumerable: false,
        configurable: true
    });
    DomainChart.prototype.s0 = function () {
        if (!this.fc) {
            this.fc = true;
            this.ta();
        }
    };
    DomainChart.prototype.ta = function () {
        this.tb();
        this.tc();
    };
    DomainChart.prototype.up = function (a) {
        if (this.ft && this.cb != null) {
            this.ft = false;
            if (a != null) {
                a.detachDataLegend(this.cb);
            }
        }
        this.uw(a);
    };
    DomainChart.prototype.st = function (a, b) {
        var c = typeCast(ITrendLineSeries_$type, a);
        if (c != null) {
            c.trendLineBrush = ArrayExtension.getModulus$1(Brush.$, this.ah, b);
            c.trendLineThickness = this.jy;
            c.trendLineType = this.d5;
            c.trendLinePeriod = this.kt;
        }
    };
    DomainChart.prototype.cw = function (a, b) {
        if (this.cz == null || this.cz.count == 0) {
            var c = this.cx(a);
            if (this.hg == null || this.hg.count == 0) {
                return a != null && a.e6 ? c : 2;
            }
            var d = ArrayExtension.getModulus$1(MarkerType_$type, this.hg, b);
            if (d == 0 && a != null) {
                d = a.e6 ? c : 2;
                if (d == 2) {
                    d = this.cv(a);
                }
            }
            if (d == 2) {
                if (a == null) {
                    return 1;
                }
                else {
                    return 2;
                }
            }
            return d;
        }
        return ArrayExtension.getModulus$1(MarkerType_$type, this.cz, b);
    };
    DomainChart.prototype.cv = function (a) {
        return 2;
    };
    DomainChart.prototype.cx = function (a) {
        return 1;
    };
    DomainChart.prototype.k4 = function () {
        if (this.itemsSource != null) {
            return this.itemsSource;
        }
        if (this.g3) {
            return this.hm();
        }
        return null;
    };
    DomainChart.prototype.get_gs = function () {
        if (this.itemsSource == null && this.g3) {
            if (this.hm() != null) {
                return true;
            }
        }
        return false;
    };
    Object.defineProperty(DomainChart.prototype, "gs", {
        get: function () {
            return this.get_gs();
        },
        enumerable: false,
        configurable: true
    });
    DomainChart.prototype.hm = function () {
        return null;
    };
    DomainChart.prototype.t5 = function (a, b, c) {
        var _this = this;
        if (this.propertyChanged != null) {
            this.propertyChanged(this, new PropertyChangedEventArgs(a));
        }
        switch (a) {
            case "DataChart":
                this.up(typeCast(XamDataChart.$, b));
                this.td(typeCast(XamDataChart.$, c));
                break;
            case "ShouldUseSkeletonStyleForMockData":
            case "ShouldDisplayMockData":
                this.actualItemsSource = this.k4();
                this.z = this.ag(this.ab);
                this.aa = this.ag(this.af);
                break;
            case "ActualBrushes":
                this.tz(function (d) { return d.ag = _this.z; }, true, a);
                break;
            case "ActualOutlines":
                this.tz(function (d) { return d.aj = _this.aa; }, true, a);
                break;
            case "ActualItemsSource":
                if (this.g3) {
                    this.z = this.ag(this.ab);
                    this.aa = this.ag(this.af);
                }
                this.bh.itemsSource = this.actualItemsSource;
                break;
            case "ItemsSource":
                this.actualItemsSource = this.k4();
                break;
            case "IncludedProperties":
                this.bh.includedProperties = this.q;
                break;
            case "ExcludedProperties":
                this.bh.excludedProperties = this.o;
                break;
            case "Brushes":
                this.z = this.ag(this.ab);
                break;
            case "Outlines":
                this.aa = this.ag(this.af);
                break;
            case "Legend":
                this.tx(function (d) { return d.legend = _this.legend; });
                break;
            case "HighlightingMode":
                this.tx(function (d) { return d.by = _this.de; });
                break;
            case "HighlightingBehavior":
                this.tx(function (d) { return d.bw = _this.dc; });
                break;
            case "LegendHighlightingMode":
                this.tx(function (d) { return d.bd = _this.ch; });
                break;
            case "LegendItemBadgeShape":
                this.tx(function (d) { return d.bf = _this.cn; });
                break;
            case "LegendItemBadgeMode":
                this.tx(function (d) { return d.be = _this.cj; });
                break;
            case "PlotAreaMarginLeft":
                this.tx(function (d) { return d.xx = _this.jj; });
                break;
            case "PlotAreaMarginTop":
                this.tx(function (d) { return d.xz = _this.jl; });
                break;
            case "PlotAreaMarginRight":
                this.tx(function (d) { return d.xy = _this.jk; });
                break;
            case "PlotAreaMarginBottom":
                this.tx(function (d) { return d.xw = _this.ji; });
                break;
            case "ComputedPlotAreaMarginMode":
                this.tx(function (d) { return d.v6 = _this.bb(); });
                break;
            case "IsHorizontalZoomEnabled":
                this.tx(function (d) { return d.ws = _this.gt; });
                break;
            case "IsVerticalZoomEnabled":
                this.tx(function (d) { return d.wv = _this.g0; });
                break;
            case "WindowRect":
                this.ty(function (d) { return d.vx = _this.ww; }, false);
                break;
            case "WindowRectMinWidth":
                this.ty(function (d) { return d.ij = _this.j8; }, false);
                break;
            case "WindowRectMinHeight":
                this.ty(function (d) { return d.ii = _this.j7; }, false);
                break;
            case "WindowSizeMinWidth":
                this.ty(function (d) { return d.il = _this.ka; }, false);
                break;
            case "WindowSizeMinHeight":
                this.ty(function (d) { return d.ik = _this.j9; }, false);
                break;
            case "Title":
                this.ty(function (d) { return d.lo = _this.rj; }, false);
                break;
            case "Subtitle":
                this.ty(function (d) { return d.lc = _this.q8; }, false);
                break;
            case "TitleAlignment":
                this.tx(function (d) { return d.ue = _this.u7; });
                break;
            case "SubtitleAlignment":
                this.tx(function (d) { return d.ud = _this.u6; });
                break;
            case "UnknownValuePlotting":
                this.s3(function (d) {
                    if (typeCast(IUnknownValuePlotting_$type, d) !== null) {
                        d.unknownValuePlotting = _this.d7;
                    }
                });
                break;
            case "IsSeriesHighlightingEnabled":
                this.s3(function (d) { return d.ey = _this.gv; });
                break;
            case "HighlightedValuesDisplayMode":
                this.s7(function (d) { return d.bm = _this.da; });
                break;
            case "Thickness":
                if (!isNaN_(this.js)) {
                    this.s7(function (d) { return d.i5 = _this.js; });
                }
                break;
            case "OutlineMode":
                this.s7(function (d) { return d.b8 = _this.dg; });
                break;
            case "MarkerFillOpacity":
                this.s7(function (d) { return d.iw = _this.jf; });
                break;
            case "MarkerThickness":
                if (!isNaN_(this.jg)) {
                    this.s7(function (d) { return d.rf(_this.jg); });
                }
                break;
            case "MarkerOutlineMode":
                this.s7(function (d) { return d.re(_this.ct); });
                break;
            case "MarkerFillMode":
                this.s7(function (d) { return d.rd(_this.cr); });
                break;
            case "Resolution":
                if (!isNaN_(this.jm)) {
                    this.s3(function (d) { return d.iz = _this.jm; });
                }
                break;
            case "AreaFillOpacity":
                if (!isNaN_(this.i0)) {
                    this.s3(function (d) { return d.h0 = _this.i0; });
                }
                break;
            case "ShouldAvoidAxisAnnotationCollisions":
                this.s2(function (d) {
                    d.cu = _this.g2;
                    d.cw = _this.g2;
                }, true);
                break;
            case "ShouldPanOnMaximumZoom":
                this.tx(function (d) { return d.fy = _this.g4; });
                break;
            case "TrendLineBrushes":
            case "TrendLineType":
            case "TrendLineThickness":
            case "TrendLinePeriod":
                this.s4(function (d, e) { return _this.st(d, e); });
                break;
            case "MarkerMaxCount":
                this.s8(function (d, e) { return _this.ss(d, e); });
                break;
            case "MarkerTypes":
                var d = typeCast(ObservableCollection$1.$.specialize(MarkerType_$type), b);
                if (d != null) {
                    d.collectionChanged = delegateRemove(d.collectionChanged, runOn(this, this.t3));
                }
                var e = typeCast(ObservableCollection$1.$.specialize(MarkerType_$type), c);
                if (e != null) {
                    e.collectionChanged = delegateCombine(e.collectionChanged, runOn(this, this.t3));
                }
                this.s8(function (f, g) { return _this.ss(f, g); });
                break;
            case "MarkerAutomaticBehavior":
                this.tx(function (f) { return f.bg = _this.cp; });
                this.s8(function (f, g) { return _this.ss(f, g); });
                break;
            case "MarkerBrushes":
                this.tx(function (f) { return f.ah = _this.ad; });
                break;
            case "MarkerOutlines":
                this.tx(function (f) { return f.ai = _this.ae; });
                break;
            case "AlignsGridLinesToPixels":
                this.tx(function (f) { return f.wo = _this.f6; });
                break;
            case "AnimateSeriesWhenAxisRangeChanges":
                this.tx(function (f) { return f.eg = _this.f7; });
                break;
            case "ToolTipType":
            case "CrosshairsDisplayMode":
            case "CrosshairsSnapToData":
            case "CrosshairsLineVerticalStroke":
            case "CrosshairsLineHorizontalStroke":
            case "CrosshairsLineThickness":
            case "CrosshairsAnnotationXAxisBackground":
            case "CrosshairsAnnotationXAxisTextColor":
            case "CrosshairsAnnotationYAxisBackground":
            case "CrosshairsAnnotationYAxisTextColor":
            case "CrosshairsAnnotationEnabled":
            case "CrosshairsAnnotationXAxisPrecision":
            case "CrosshairsAnnotationYAxisPrecision":
            case "FinalValueAnnotationsVisible":
            case "FinalValueAnnotationsPrecision":
            case "FinalValueAnnotationsTextColor":
            case "FinalValueAnnotationsBackground":
            case "CalloutsVisible":
            case "AutoCalloutsVisible":
            case "CalloutsAllowedPositions":
            case "CalloutsItemsSource":
            case "CalloutsXMemberPath":
            case "CalloutsYMemberPath":
            case "CalloutsLabelMemberPath":
            case "CalloutsContentMemberPath":
                var f = this.uc(runOn(this, this.sz), this.gu);
                this.gu = f.p1;
                break;
            case "DataToolTipGroupingMode":
            case "DataToolTipPositionOffsetX":
            case "DataToolTipPositionOffsetY":
            case "DataToolTipDefaultPositionOffsetX":
            case "DataToolTipDefaultPositionOffsetY":
            case "DataToolTipGroupedPositionModeX":
            case "DataToolTipGroupedPositionModeY":
            case DomainChart.nx:
            case "DataToolTipIncludedSeries":
            case "DataToolTipExcludedSeries":
            case "DataToolTipIncludedColumns":
            case "DataToolTipExcludedColumns":
            case "DataToolTipValueFormatAbbreviation":
            case "DataToolTipValueFormatMaxFractions":
            case "DataToolTipValueFormatMinFractions":
            case "DataToolTipValueFormatMode":
            case "DataToolTipValueFormatCulture":
            case "DataToolTipValueFormatUseGrouping":
            case "DataToolTipValueFormatString":
            case "DataToolTipValueFormatSpecifiers":
            case "DataToolTipValueRowMargin":
            case "DataToolTipValueRowVisible":
            case "DataToolTipValueTextWhenMissingData":
            case "DataToolTipValueTextUseSeriesColors":
            case "DataToolTipValueTextMargin":
            case "DataToolTipValueTextColor":
            case "DataToolTipValueTextStyle":
            case "DataToolTipHeaderFormatString":
            case "DataToolTipHeaderFormatSpecifiers":
            case "DataToolTipHeaderFormatCulture":
            case "DataToolTipHeaderFormatDate":
            case "DataToolTipHeaderFormatTime":
            case "DataToolTipHeaderText":
            case "DataToolTipHeaderTextColor":
            case "DataToolTipHeaderTextMargin":
            case "DataToolTipHeaderRowMargin":
            case "DataToolTipHeaderRowVisible":
            case "DataToolTipHeaderTextStyle":
            case "DataToolTipGroupTextColor":
            case "DataToolTipGroupTextMargin":
            case "DataToolTipGroupRowMargin":
            case "DataToolTipGroupRowVisible":
            case "DataToolTipGroupTextStyle":
            case "DataToolTipSummaryTitleTextColor":
            case "DataToolTipSummaryTitleTextStyle":
            case "DataToolTipSummaryType":
            case "DataToolTipSummaryTitleText":
            case "DataToolTipSummaryTitleTextMargin":
            case "DataToolTipSummaryRowMargin":
            case "DataToolTipSummaryValueTextColor":
            case "DataToolTipSummaryValueTextStyle":
            case "DataToolTipSummaryLabelText":
            case "DataToolTipSummaryLabelTextColor":
            case "DataToolTipSummaryLabelTextStyle":
            case "DataToolTipSummaryUnitsText":
            case "DataToolTipSummaryUnitsTextColor":
            case "DataToolTipSummaryUnitsTextStyle":
            case "DataToolTipBadgeMargin":
            case "DataToolTipBadgeShape":
            case "DataToolTipUnitsDisplayMode":
            case "DataToolTipUnitsText":
            case "DataToolTipUnitsTextColor":
            case "DataToolTipUnitsTextMargin":
            case "DataToolTipUnitsTextStyle":
            case "DataToolTipTitleTextMargin":
            case "DataToolTipTitleTextColor":
            case "DataToolTipTitleTextStyle":
            case "DataToolTipLabelDisplayMode":
            case "DataToolTipLabelTextColor":
            case "DataToolTipLabelTextMargin":
            case "DataToolTipLabelTextStyle":
                this.t1(a, b, c);
                break;
            case "SelectionMode":
                this.ty(function (g) { return g.cb = _this.dr; }, false);
                break;
            case "FocusMode":
                this.ty(function (g) { return g.b9 = _this.dq; }, false);
                break;
            case "SelectionBehavior":
                this.ty(function (g) { return g.b6 = _this.dm; }, false);
                break;
            case "SelectionBrush":
                this.ty(function (g) { return g.u1 = _this.wo; }, false);
                break;
            case "FocusBrush":
                this.ty(function (g) { return g.uk = _this.wg; }, false);
                break;
            case "HorizontalViewScrollbarMode":
                this.tx(function (g) { return g.ch = _this.dw; });
                break;
            case "VerticalViewScrollbarMode":
                this.tx(function (g) { return g.ci = _this.dx; });
                break;
            case "HorizontalViewScrollbarCornerRadius":
                this.tx(function (g) { return g.hl = _this.i7; });
                break;
            case "HorizontalViewScrollbarFill":
                this.tx(function (g) { return g.uv = _this.wl; });
                break;
            case "HorizontalViewScrollbarMaxOpacity":
                this.tx(function (g) { return g.ho = _this.ja; });
                break;
            case "HorizontalViewScrollbarOutline":
                this.tx(function (g) { return g.uw = _this.wm; });
                break;
            case "HorizontalViewScrollbarStrokeThickness":
                this.tx(function (g) { return g.hp = _this.jb; });
                break;
            case "VerticalViewScrollbarCornerRadius":
                this.tx(function (g) { return g.h8 = _this.j0; });
                break;
            case "VerticalViewScrollbarFill":
                this.tx(function (g) { return g.u5 = _this.wr; });
                break;
            case "VerticalViewScrollbarMaxOpacity":
                this.tx(function (g) { return g.ia = _this.j2; });
                break;
            case "VerticalViewScrollbarOutline":
                this.tx(function (g) { return g.u6 = _this.ws; });
                break;
            case "VerticalViewScrollbarStrokeThickness":
                this.tx(function (g) { return g.ib = _this.j3; });
                break;
            case "HorizontalViewScrollbarHeight":
                this.tx(function (g) { return g.hm = _this.i8; });
                break;
            case "HorizontalViewScrollbarInset":
                this.tx(function (g) { return g.hn = _this.i9; });
                break;
            case DomainChart.pt:
                this.tx(function (g) { return g.et = _this.gm; });
                break;
            case "HorizontalViewScrollbarTrackStartInset":
                this.tx(function (g) { return g.hr = _this.jd; });
                break;
            case "HorizontalViewScrollbarTrackEndInset":
                this.tx(function (g) { return g.hq = _this.jc; });
                break;
            case "VerticalViewScrollbarPosition":
                this.tx(function (g) { return g.cl = _this.d1; });
                break;
            case "HorizontalViewScrollbarPosition":
                this.tx(function (g) { return g.ce = _this.dt; });
                break;
            case "ValueLines":
                var g = typeCast(ObservableCollection$1.$.specialize(ValueLayerValueMode_$type), b);
                if (g != null) {
                    g.collectionChanged = delegateRemove(g.collectionChanged, runOn(this, this.t7));
                }
                var h = typeCast(ObservableCollection$1.$.specialize(ValueLayerValueMode_$type), c);
                if (h != null) {
                    h.collectionChanged = delegateCombine(h.collectionChanged, runOn(this, this.t7));
                }
                var i = this.uc(runOn(this, this.sz), this.gu);
                this.gu = i.p1;
                break;
        }
        this.t8(a, b, c);
        this.t9(a, b, c);
        if (this.gx) {
            this.cd.onTargetPropertyChanged(this, a, b, c);
        }
    };
    DomainChart.prototype.t1 = function (a, b, c) {
        this.sr(a, c);
    };
    DomainChart.prototype.sr = function (a, b) {
        var _this = this;
        if (!this.g8.containsKey(a)) {
            this.ha.add(a);
            this.g8.addItem(a, b);
        }
        else {
            this.g8.item(a, b);
        }
        this.s5(function (c) {
            if (_this.b8.isDataToolTipLayer(c)) {
                _this.sq(c);
            }
        });
    };
    DomainChart.prototype.sq = function (a) {
        for (var b = 0; b < this.ha.count; b++) {
            var c = this.ha._inner[b];
            var d = this.g8.item(c);
            if (stringStartsWith(c, "DataToolTip")) {
                c = c.substr(("DataToolTip").length);
            }
            this.b8.updateDataToolTipPropertyByName(a, c, d);
        }
    };
    DomainChart.prototype.t3 = function (a, b) {
        var _this = this;
        this.s8(function (c, d) { return _this.ss(c, d); });
    };
    DomainChart.prototype.t7 = function (a, b) {
        var c = this.uc(runOn(this, this.sz), this.gu);
        this.gu = c.p1;
    };
    Object.defineProperty(DomainChart.prototype, "a4", {
        get: function () {
            return this.ay;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "a0", {
        get: function () {
            return this.au;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "az", {
        get: function () {
            return this.at;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "eg", {
        get: function () {
            return this.ec;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "eh", {
        get: function () {
            return this.ed;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "a8", {
        get: function () {
            return this.a6;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "dr", {
        get: function () {
            return this.dp;
        },
        set: function (a) {
            var b = this.dp;
            this.dp = a;
            if (b != this.dp) {
                this.t5("SelectionMode", enumGetBox(SeriesSelectionMode_$type, b), enumGetBox(SeriesSelectionMode_$type, this.dp));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "dq", {
        get: function () {
            return this.dn;
        },
        set: function (a) {
            var b = this.dn;
            this.dn = a;
            if (b != this.dn) {
                this.t5("FocusMode", enumGetBox(SeriesSelectionMode_$type, b), enumGetBox(SeriesSelectionMode_$type, this.dn));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "wo", {
        get: function () {
            return this.vt;
        },
        set: function (a) {
            var b = this.vt;
            this.vt = a;
            if (b != this.vt) {
                this.t5("SelectionBrush", b, this.vt);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "wg", {
        get: function () {
            return this.vq;
        },
        set: function (a) {
            var b = this.vq;
            this.vq = a;
            if (b != this.vq) {
                this.t5("FocusBrush", b, this.vq);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "dm", {
        get: function () {
            return this.dl;
        },
        set: function (a) {
            var b = this.dl;
            this.dl = a;
            if (b != this.dl) {
                this.t5("SelectionBehavior", enumGetBox(SeriesSelectionBehavior_$type, b), enumGetBox(SeriesSelectionBehavior_$type, this.dl));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "selectedSeriesItems", {
        get: function () {
            if (this.dataChart != null) {
                return this.dataChart.selectedSeriesItems;
            }
            return this.ao;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "focusedSeriesItems", {
        get: function () {
            if (this.dataChart != null) {
                return this.dataChart.focusedSeriesItems;
            }
            return this.an;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "a3", {
        get: function () {
            return this.ax;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "a2", {
        get: function () {
            return this.aw;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "a1", {
        get: function () {
            return this.av;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "ei", {
        get: function () {
            return this.ee;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "ej", {
        get: function () {
            return this.ef;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "a7", {
        get: function () {
            return this.a5;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "p4", {
        get: function () {
            return this.ls;
        },
        set: function (a) {
            var _this = this;
            var b = this.ls;
            this.ls = a;
            if (b != a) {
                this.ax.n();
                this.ek.b(this.ls, function (c, d, e) { return _this.ax.j.add(new ChartSortDescription(2, c, d)); });
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "pg", {
        get: function () {
            return this.lo;
        },
        set: function (a) {
            var _this = this;
            var b = this.lo;
            this.lo = a;
            if (b != a) {
                this.aw.n();
                this.ek.b(this.lo, function (c, d, e) { return _this.aw.j.add(new ChartSortDescription(2, c, d)); });
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "p0", {
        get: function () {
            return this.lq;
        },
        set: function (a) {
            var _this = this;
            var b = this.lq;
            this.lq = a;
            if (b != a) {
                this.av.n();
                this.ek.b(this.lq, function (c, d, e) { return _this.av.j.add(new ChartSortDescription(2, c, d)); });
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "py", {
        get: function () {
            return this.lp;
        },
        set: function (a) {
            var b = this.lp;
            this.lp = a;
            if (b != a) {
                this.ee.clear();
                var c_1 = FilterExpressionParsingHelper.getFilterExpression(a);
                if (!stringIsNullOrEmpty(c_1.b)) {
                    if (this.filterStringErrorsParsing != null) {
                        this.filterStringErrorsParsing(this, ((function () {
                            var $ret = new FilterStringErrorsParsingEventArgs();
                            $ret.propertyName = "InitialFilter";
                            $ret.errors = c_1.b;
                            return $ret;
                        })()));
                    }
                    return;
                }
                this.ee.k.add(c_1.a);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "p2", {
        get: function () {
            return this.lr;
        },
        set: function (a) {
            var b = this.lr;
            this.lr = a;
            if (b != a) {
                this.ef.clear();
                var c_2 = FilterExpressionParsingHelper.getFilterExpression(a);
                if (!stringIsNullOrEmpty(c_2.b)) {
                    if (this.filterStringErrorsParsing != null) {
                        this.filterStringErrorsParsing(this, ((function () {
                            var $ret = new FilterStringErrorsParsingEventArgs();
                            $ret.propertyName = "InitialFilter";
                            $ret.errors = c_2.b;
                            return $ret;
                        })()));
                    }
                    return;
                }
                this.ef.k.add(c_2.a);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "p6", {
        get: function () {
            return this.lt;
        },
        set: function (a) {
            var _this = this;
            var b = this.lt;
            this.lt = a;
            if (b != a) {
                this.a5.n();
                this.el.a(this.lt, function (c, d, e) {
                    return _this.a5.j.add(((function () {
                        var $ret = new ChartSummaryDescription(2, c, d);
                        $ret.n = e;
                        return $ret;
                    })()));
                });
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "itemsSource", {
        get: function () {
            return this.kv;
        },
        set: function (a) {
            var b = this.itemsSource;
            if (b != a) {
                this.kv = a;
                this.t5("ItemsSource", b, this.itemsSource);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "actualItemsSource", {
        get: function () {
            return this.ku;
        },
        set: function (a) {
            var b = this.actualItemsSource;
            if (b != a) {
                this.ku = a;
                this.t5("ActualItemsSource", b, this.actualItemsSource);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "q", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            var b = this.h;
            this.h = a;
            if (b != this.h) {
                this.t5("IncludedProperties", b, this.h);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "o", {
        get: function () {
            return this.g;
        },
        set: function (a) {
            var b = this.g;
            this.g = a;
            if (b != this.g) {
                this.t5("ExcludedProperties", b, this.g);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "ab", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            var b = this.ab;
            if (a != b) {
                this.t = a;
                this.t5("Brushes", b, this.ab);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "af", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            var b = this.af;
            if (a != b) {
                this.w = a;
                this.t5("Outlines", b, this.af);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "z", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            var b = this.r;
            if (a != b) {
                this.r = a;
                this.t5("ActualBrushes", b, this.z);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "aa", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            var b = this.s;
            if (a != b) {
                this.s = a;
                this.t5("ActualOutlines", b, this.aa);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "legend", {
        get: function () {
            return this.b9;
        },
        set: function (a) {
            var b = this.legend;
            if (a != b) {
                this.b9 = a;
                this.t5("Legend", b, this.legend);
            }
        },
        enumerable: false,
        configurable: true
    });
    DomainChart.prototype.get_gt = function () {
        return this.fs;
    };
    DomainChart.prototype.set_gt = function (a) {
        var b = this.gt;
        if (a != b) {
            this.fs = a;
            this.t5("IsHorizontalZoomEnabled", b, this.gt);
        }
    };
    Object.defineProperty(DomainChart.prototype, "gt", {
        get: function () {
            return this.get_gt();
        },
        set: function (a) {
            this.set_gt(a);
        },
        enumerable: false,
        configurable: true
    });
    DomainChart.prototype.get_g0 = function () {
        return this.fv;
    };
    DomainChart.prototype.set_g0 = function (a) {
        var b = this.g0;
        if (a != b) {
            this.fv = a;
            this.t5("IsVerticalZoomEnabled", b, this.g0);
        }
    };
    Object.defineProperty(DomainChart.prototype, "g0", {
        get: function () {
            return this.get_g0();
        },
        set: function (a) {
            this.set_g0(a);
        },
        enumerable: false,
        configurable: true
    });
    DomainChart.prototype.get_g3 = function () {
        return this.fx;
    };
    DomainChart.prototype.set_g3 = function (a) {
        var b = this.g3;
        if (a != b) {
            this.fx = a;
            this.t5("ShouldDisplayMockData", b, this.g3);
        }
    };
    Object.defineProperty(DomainChart.prototype, "g3", {
        get: function () {
            return this.get_g3();
        },
        set: function (a) {
            this.set_g3(a);
        },
        enumerable: false,
        configurable: true
    });
    DomainChart.prototype.get_g6 = function () {
        return this.f4;
    };
    DomainChart.prototype.set_g6 = function (a) {
        var b = this.g6;
        if (a != b) {
            this.f4 = a;
            this.t5("ShouldUseSkeletonStyleForMockData", b, this.g6);
        }
    };
    Object.defineProperty(DomainChart.prototype, "g6", {
        get: function () {
            return this.get_g6();
        },
        set: function (a) {
            this.set_g6(a);
        },
        enumerable: false,
        configurable: true
    });
    DomainChart.prototype.get_da = function () {
        return this.c9;
    };
    DomainChart.prototype.set_da = function (a) {
        var b = this.da;
        if (a != b) {
            this.c9 = a;
            this.t5("HighlightedValuesDisplayMode", enumGetBox(SeriesHighlightedValuesDisplayMode_$type, b), enumGetBox(SeriesHighlightedValuesDisplayMode_$type, this.da));
        }
    };
    Object.defineProperty(DomainChart.prototype, "da", {
        get: function () {
            return this.get_da();
        },
        set: function (a) {
            this.set_da(a);
        },
        enumerable: false,
        configurable: true
    });
    DomainChart.prototype.get_gv = function () {
        return this.fu;
    };
    DomainChart.prototype.set_gv = function (a) {
        var b = this.gv;
        if (a != b) {
            this.fu = a;
            this.t5("IsSeriesHighlightingEnabled", b, this.gv);
        }
    };
    Object.defineProperty(DomainChart.prototype, "gv", {
        get: function () {
            return this.get_gv();
        },
        set: function (a) {
            this.set_gv(a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "ww", {
        get: function () {
            return this.wu;
        },
        set: function (a) {
            var b = this.ww;
            if (Rect.l_op_Inequality(a, b)) {
                this.wu = a;
                this.t5("WindowRect", b, this.ww);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "wv", {
        get: function () {
            return this.dataChart != null ? this.dataChart.vw : Rect.empty;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "rj", {
        get: function () {
            return this.lw;
        },
        set: function (a) {
            var b = this.rj;
            if (a != b) {
                this.lw = a;
                this.t5("Title", b, this.rj);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "q8", {
        get: function () {
            return this.lu;
        },
        set: function (a) {
            var b = this.q8;
            if (a != b) {
                this.lu = a;
                this.t5("Subtitle", b, this.q8);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "u7", {
        get: function () {
            return this.u5;
        },
        set: function (a) {
            var b = this.u7;
            if (b != a) {
                this.u5 = a;
                this.t5("TitleAlignment", enumGetBox(HorizontalAlignment_$type, b), enumGetBox(HorizontalAlignment_$type, this.u7));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "u6", {
        get: function () {
            return this.u4;
        },
        set: function (a) {
            var b = this.u6;
            if (b != a) {
                this.u4 = a;
                this.t5("SubtitleAlignment", enumGetBox(HorizontalAlignment_$type, b), enumGetBox(HorizontalAlignment_$type, this.u6));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "d7", {
        get: function () {
            return this.d6;
        },
        set: function (a) {
            var b = this.d7;
            if (a != b) {
                this.d6 = a;
                this.t5("UnknownValuePlotting", enumGetBox(UnknownValuePlotting_$type, b), enumGetBox(UnknownValuePlotting_$type, this.d7));
            }
        },
        enumerable: false,
        configurable: true
    });
    DomainChart.prototype.get_jm = function () {
        return this.h9;
    };
    DomainChart.prototype.set_jm = function (a) {
        var b = this.jm;
        if (a != b) {
            this.h9 = a;
            this.t5("Resolution", b, this.jm);
        }
    };
    Object.defineProperty(DomainChart.prototype, "jm", {
        get: function () {
            return this.get_jm();
        },
        set: function (a) {
            this.set_jm(a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "js", {
        get: function () {
            return this.ig;
        },
        set: function (a) {
            var b = this.js;
            if (a != b) {
                this.ig = a;
                this.t5("Thickness", b, this.js);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "dg", {
        get: function () {
            return this.df;
        },
        set: function (a) {
            var b = this.dg;
            if (a != b) {
                this.df = a;
                this.t5("OutlineMode", enumGetBox(SeriesOutlineMode_$type, b), enumGetBox(SeriesOutlineMode_$type, this.dg));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "ct", {
        get: function () {
            return this.cs;
        },
        set: function (a) {
            var b = this.ct;
            if (a != b) {
                this.cs = a;
                this.t5("MarkerOutlineMode", enumGetBox(MarkerOutlineMode_$type, b), enumGetBox(MarkerOutlineMode_$type, this.ct));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "cr", {
        get: function () {
            return this.cq;
        },
        set: function (a) {
            var b = this.cr;
            if (a != b) {
                this.cq = a;
                this.t5("MarkerFillMode", enumGetBox(MarkerFillMode_$type, b), enumGetBox(MarkerFillMode_$type, this.cr));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "jf", {
        get: function () {
            return this.h2;
        },
        set: function (a) {
            var b = this.jf;
            if (a != b) {
                this.h2 = a;
                this.t5("MarkerFillOpacity", b, this.jf);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "jg", {
        get: function () {
            return this.h3;
        },
        set: function (a) {
            var b = this.jg;
            if (a != b) {
                this.h3 = a;
                this.t5("MarkerThickness", b, this.jg);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "cz", {
        get: function () {
            return this.cy;
        },
        set: function (a) {
            var b = this.cz;
            if (a != b) {
                this.cy = a;
                this.t5("MarkerTypes", b, this.cz);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "cp", {
        get: function () {
            return this.co;
        },
        set: function (a) {
            var b = this.cp;
            if (a != b) {
                this.co = a;
                this.t5("MarkerAutomaticBehavior", enumGetBox(MarkerAutomaticBehavior_$type, b), enumGetBox(MarkerAutomaticBehavior_$type, this.cp));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "ad", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            var b = this.ad;
            if (a != b) {
                this.u = a;
                this.t5("MarkerBrushes", b, this.ad);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "ae", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            var b = this.ae;
            if (!Base.equalsStatic(a, b)) {
                this.v = a;
                this.t5("MarkerOutlines", b, this.ae);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "kr", {
        get: function () {
            return this.ki;
        },
        set: function (a) {
            var b = this.ki;
            if (a != b) {
                this.ki = a;
                this.t5("MarkerMaxCount", b, this.ki);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "i0", {
        get: function () {
            return this.hn;
        },
        set: function (a) {
            var b = this.i0;
            if (a != b) {
                this.hn = a;
                this.t5("AreaFillOpacity", b, this.i0);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "f7", {
        get: function () {
            return this.fb;
        },
        set: function (a) {
            var b = this.fb;
            if (a != b) {
                this.fb = a;
                this.t5("AnimateSeriesWhenAxisRangeChanges", b, this.fb);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "ah", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            var b = this.ah;
            if (a != b) {
                this.x = a;
                this.t5("TrendLineBrushes", b, this.ah);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "d5", {
        get: function () {
            return this.d4;
        },
        set: function (a) {
            var b = this.d5;
            if (a != b) {
                this.d4 = a;
                this.t5("TrendLineType", enumGetBox(TrendLineType_$type, b), enumGetBox(TrendLineType_$type, this.d5));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "jy", {
        get: function () {
            return this.im;
        },
        set: function (a) {
            var b = this.jy;
            if (a != b) {
                this.im = a;
                this.t5("TrendLineThickness", b, this.jy);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "f6", {
        get: function () {
            return this.fa;
        },
        set: function (a) {
            var b = this.f6;
            if (a != b) {
                this.fa = a;
                this.t5("AlignsGridLinesToPixels", b, this.f6);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "jj", {
        get: function () {
            return this.h6;
        },
        set: function (a) {
            var b = this.h6;
            this.h6 = a;
            if (a != b) {
                this.t5("PlotAreaMarginLeft", b, a);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "jl", {
        get: function () {
            return this.h8;
        },
        set: function (a) {
            var b = this.h8;
            this.h8 = a;
            if (a != b) {
                this.t5("PlotAreaMarginTop", b, a);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "jk", {
        get: function () {
            return this.h7;
        },
        set: function (a) {
            var b = this.h7;
            this.h7 = a;
            if (a != b) {
                this.t5("PlotAreaMarginRight", b, a);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "ji", {
        get: function () {
            return this.h5;
        },
        set: function (a) {
            var b = this.h5;
            this.h5 = a;
            if (a != b) {
                this.t5("PlotAreaMarginBottom", b, a);
            }
        },
        enumerable: false,
        configurable: true
    });
    DomainChart.prototype.t4 = function () {
    };
    Object.defineProperty(DomainChart.prototype, "ba", {
        get: function () {
            return this.a9;
        },
        set: function (a) {
            var b = this.ba;
            if (a != b) {
                this.a9 = a;
                this.t5("ComputedPlotAreaMarginMode", enumGetBox(ComputedPlotAreaMarginMode_$type, b), enumGetBox(ComputedPlotAreaMarginMode_$type, this.ba));
            }
        },
        enumerable: false,
        configurable: true
    });
    DomainChart.prototype.bb = function () {
        return this.ba;
    };
    Object.defineProperty(DomainChart.prototype, "di", {
        get: function () {
            return this.dh;
        },
        set: function (a) {
            var b = this.di;
            if (a != b) {
                this.dh = a;
                this.t5("SeriesPlotAreaMarginHorizontalMode", enumGetBox(SeriesPlotAreaMarginHorizontalMode_$type, b), enumGetBox(SeriesPlotAreaMarginHorizontalMode_$type, this.di));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "dk", {
        get: function () {
            return this.dj;
        },
        set: function (a) {
            var b = this.dk;
            if (a != b) {
                this.dj = a;
                this.t5("SeriesPlotAreaMarginVerticalMode", enumGetBox(SeriesPlotAreaMarginVerticalMode_$type, b), enumGetBox(SeriesPlotAreaMarginVerticalMode_$type, this.dk));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "de", {
        get: function () {
            return this.dd;
        },
        set: function (a) {
            var b = this.de;
            if (a != b) {
                this.dd = a;
                this.t5("HighlightingMode", enumGetBox(SeriesHighlightingMode_$type, b), enumGetBox(SeriesHighlightingMode_$type, this.de));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "dc", {
        get: function () {
            return this.db;
        },
        set: function (a) {
            var b = this.dc;
            if (a != b) {
                this.db = a;
                this.t5("HighlightingBehavior", enumGetBox(SeriesHighlightingBehavior_$type, b), enumGetBox(SeriesHighlightingBehavior_$type, this.dc));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "ch", {
        get: function () {
            return this.cg;
        },
        set: function (a) {
            var b = this.ch;
            if (a != b) {
                this.cg = a;
                this.t5("LegendHighlightingMode", enumGetBox(LegendHighlightingMode_$type, b), enumGetBox(LegendHighlightingMode_$type, this.ch));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "cn", {
        get: function () {
            return this.cl;
        },
        set: function (a) {
            var b = this.cn;
            if (a != b) {
                this.cl = a;
                this.t5("LegendItemBadgeShape", enumGetBox(LegendItemBadgeShape_$type, b), enumGetBox(LegendItemBadgeShape_$type, this.cn));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "cj", {
        get: function () {
            return this.ci;
        },
        set: function (a) {
            var b = this.cj;
            if (a != b) {
                this.ci = a;
                this.t5("LegendItemBadgeMode", enumGetBox(LegendItemBadgeMode_$type, b), enumGetBox(LegendItemBadgeMode_$type, this.cj));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "kt", {
        get: function () {
            return this.kl;
        },
        set: function (a) {
            var b = this.kt;
            if (a != b) {
                this.kl = a;
                this.t5("TrendLinePeriod", b, this.kt);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "d3", {
        get: function () {
            return this.d2;
        },
        set: function (a) {
            var b = this.d3;
            if (a != b) {
                this.d2 = a;
                this.t5("ToolTipType", enumGetBox(ToolTipType_$type, b), enumGetBox(ToolTipType_$type, this.d3));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "bd", {
        get: function () {
            return this.bc;
        },
        set: function (a) {
            var b = this.bd;
            if (a != b) {
                this.bc = a;
                this.t5("CrosshairsDisplayMode", enumGetBox(CrosshairsDisplayMode_$type, b), enumGetBox(CrosshairsDisplayMode_$type, this.bd));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "gd", {
        get: function () {
            return this.fh;
        },
        set: function (a) {
            var b = this.gd;
            if (a != b) {
                this.fh = a;
                this.t5("CrosshairsSnapToData", b, this.gd);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "v3", {
        get: function () {
            return this.vd;
        },
        set: function (a) {
            var b = this.v3;
            if (a != b) {
                this.vd = a;
                this.t5("CrosshairsLineVerticalStroke", b, this.v3);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "i2", {
        get: function () {
            return this.hp;
        },
        set: function (a) {
            var b = this.i2;
            if (a != b) {
                this.hp = a;
                this.t5("CrosshairsLineThickness", b, this.i2);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "v2", {
        get: function () {
            return this.vc;
        },
        set: function (a) {
            var b = this.v2;
            if (a != b) {
                this.vc = a;
                this.t5("CrosshairsLineHorizontalStroke", b, this.v2);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "gc", {
        get: function () {
            return this.fg;
        },
        set: function (a) {
            var b = this.gc;
            if (a != b) {
                this.fg = a;
                this.t5("CrosshairsAnnotationEnabled", b, this.gc);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "vy", {
        get: function () {
            return this.u8;
        },
        set: function (a) {
            var b = this.vy;
            if (a != b) {
                this.u8 = a;
                this.t5("CrosshairsAnnotationXAxisBackground", b, this.vy);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "v0", {
        get: function () {
            return this.va;
        },
        set: function (a) {
            var b = this.v0;
            if (a != b) {
                this.va = a;
                this.t5("CrosshairsAnnotationYAxisBackground", b, this.v0);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "vz", {
        get: function () {
            return this.u9;
        },
        set: function (a) {
            var b = this.vz;
            if (a != b) {
                this.u9 = a;
                this.t5("CrosshairsAnnotationXAxisTextColor", b, this.vz);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "v1", {
        get: function () {
            return this.vb;
        },
        set: function (a) {
            var b = this.v1;
            if (a != b) {
                this.vb = a;
                this.t5("CrosshairsAnnotationYAxisTextColor", b, this.v1);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "km", {
        get: function () {
            return this.kd;
        },
        set: function (a) {
            var b = this.km;
            if (a != b) {
                this.kd = a;
                this.t5("CrosshairsAnnotationXAxisPrecision", b, this.km);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "kn", {
        get: function () {
            return this.ke;
        },
        set: function (a) {
            var b = this.kn;
            if (a != b) {
                this.ke = a;
                this.t5("CrosshairsAnnotationYAxisPrecision", b, this.kn);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "g2", {
        get: function () {
            return this.fw;
        },
        set: function (a) {
            var b = this.g2;
            if (a != b) {
                this.fw = a;
                this.t5("ShouldAvoidAxisAnnotationCollisions", b, this.g2);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "g4", {
        get: function () {
            return this.fy;
        },
        set: function (a) {
            var b = this.fy;
            if (a != b) {
                this.fy = a;
                this.t5("ShouldPanOnMaximumZoom", b, this.fy);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "gk", {
        get: function () {
            return this.fo;
        },
        set: function (a) {
            var b = this.gk;
            if (a != b) {
                this.fo = a;
                this.t5("FinalValueAnnotationsVisible", b, this.gk);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "we", {
        get: function () {
            return this.vo;
        },
        set: function (a) {
            var b = this.we;
            if (a != b) {
                this.vo = a;
                this.t5("FinalValueAnnotationsBackground", b, this.we);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "wf", {
        get: function () {
            return this.vp;
        },
        set: function (a) {
            var b = this.wf;
            if (a != b) {
                this.vp = a;
                this.t5("FinalValueAnnotationsTextColor", b, this.wf);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "kq", {
        get: function () {
            return this.kh;
        },
        set: function (a) {
            var b = this.kq;
            if (a != b) {
                this.kh = a;
                this.t5("FinalValueAnnotationsPrecision", b, this.kq);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "f9", {
        get: function () {
            return this.fd;
        },
        set: function (a) {
            var b = this.f9;
            if (a != b) {
                this.fd = a;
                this.t5("AutoCalloutsVisible", b, this.f9);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "gb", {
        get: function () {
            return this.ff;
        },
        set: function (a) {
            var b = this.gb;
            if (a != b) {
                this.ff = a;
                this.t5("CalloutsVisible", b, this.gb);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "ga", {
        get: function () {
            return this.fe;
        },
        set: function (a) {
            var b = this.ga;
            if (a != b) {
                this.fe = a;
                this.t5("CalloutStyleUpdatingEventEnabled", b, this.ga);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "calloutsItemsSource", {
        get: function () {
            return this.hk;
        },
        set: function (a) {
            var b = this.calloutsItemsSource;
            if (a != b) {
                this.hk = a;
                this.t5("CalloutsItemsSource", b, this.calloutsItemsSource);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "ak", {
        get: function () {
            return this.aj;
        },
        set: function (a) {
            var b = this.ak;
            if (a != b) {
                this.aj = a;
                this.t5("CalloutsAllowedPositions", b, this.ak);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "mf", {
        get: function () {
            return this.lc;
        },
        set: function (a) {
            var b = this.mf;
            if (a != b) {
                this.lc = a;
                this.t5("CalloutsXMemberPath", b, this.mf);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "mh", {
        get: function () {
            return this.ld;
        },
        set: function (a) {
            var b = this.mh;
            if (a != b) {
                this.ld = a;
                this.t5("CalloutsYMemberPath", b, this.mh);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "mb", {
        get: function () {
            return this.lb;
        },
        set: function (a) {
            var b = this.mb;
            if (a != b) {
                this.lb = a;
                this.t5("CalloutsLabelMemberPath", b, this.mb);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "l8", {
        get: function () {
            return this.la;
        },
        set: function (a) {
            var b = this.l8;
            if (a != b) {
                this.la = a;
                this.t5("CalloutsContentMemberPath", b, this.l8);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "d9", {
        get: function () {
            return this.d8;
        },
        set: function (a) {
            var b = this.d8;
            if (a != b) {
                this.d8 = a;
                this.t5("ValueLines", b, this.d8);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "ai", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            var b = this.y;
            if (a != b) {
                this.y = a;
                this.t5("ValueLinesBrushes", b, this.y);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "jz", {
        get: function () {
            return this.io;
        },
        set: function (a) {
            var b = this.io;
            if (a != b) {
                this.io = a;
                this.t5("ValueLinesThickness", b, this.io);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "dw", {
        get: function () {
            return this.du;
        },
        set: function (a) {
            var b = this.du;
            this.du = a;
            if (b != this.du) {
                this.t5("HorizontalViewScrollbarMode", enumGetBox(SeriesViewerScrollbarMode_$type, b), enumGetBox(SeriesViewerScrollbarMode_$type, this.du));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "dx", {
        get: function () {
            return this.dv;
        },
        set: function (a) {
            var b = this.dv;
            this.dv = a;
            if (b != this.dv) {
                this.t5("VerticalViewScrollbarMode", enumGetBox(SeriesViewerScrollbarMode_$type, b), enumGetBox(SeriesViewerScrollbarMode_$type, this.dv));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "dt", {
        get: function () {
            return this.ds;
        },
        set: function (a) {
            var b = this.ds;
            this.ds = a;
            if (b != this.ds) {
                this.t5("HorizontalViewScrollbarPosition", enumGetBox(SeriesViewerHorizontalScrollbarPosition_$type, b), enumGetBox(SeriesViewerHorizontalScrollbarPosition_$type, this.ds));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "d1", {
        get: function () {
            return this.d0;
        },
        set: function (a) {
            var b = this.d0;
            this.d0 = a;
            if (b != this.d0) {
                this.t5("VerticalViewScrollbarPosition", enumGetBox(SeriesViewerVerticalScrollbarPosition_$type, b), enumGetBox(SeriesViewerVerticalScrollbarPosition_$type, this.d0));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "wl", {
        get: function () {
            return this.vr;
        },
        set: function (a) {
            var b = this.vr;
            this.vr = a;
            if (b != this.vr) {
                this.t5("HorizontalViewScrollbarFill", b, this.vr);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "wm", {
        get: function () {
            return this.vs;
        },
        set: function (a) {
            var b = this.vs;
            this.vs = a;
            if (b != this.vs) {
                this.t5("HorizontalViewScrollbarOutline", b, this.vs);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "jb", {
        get: function () {
            return this.hy;
        },
        set: function (a) {
            var b = this.hy;
            this.hy = a;
            if (b != this.hy) {
                this.t5("HorizontalViewScrollbarStrokeThickness", b, this.hy);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "ja", {
        get: function () {
            return this.hx;
        },
        set: function (a) {
            var b = this.hx;
            this.hx = a;
            if (b != this.hx) {
                this.t5("HorizontalViewScrollbarMaxOpacity", b, this.hx);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "i7", {
        get: function () {
            return this.hu;
        },
        set: function (a) {
            var b = this.hu;
            this.hu = a;
            if (b != this.hu) {
                this.t5("HorizontalViewScrollbarCornerRadius", b, this.hu);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "i8", {
        get: function () {
            return this.hv;
        },
        set: function (a) {
            var b = this.hv;
            this.hv = a;
            if (b != this.hv) {
                this.t5("HorizontalViewScrollbarHeight", b, this.hv);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "i9", {
        get: function () {
            return this.hw;
        },
        set: function (a) {
            var b = this.hw;
            this.hw = a;
            if (b != this.hw) {
                this.t5("HorizontalViewScrollbarInset", b, this.hw);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "jd", {
        get: function () {
            return this.h0;
        },
        set: function (a) {
            var b = this.h0;
            this.h0 = a;
            if (b != this.h0) {
                this.t5("HorizontalViewScrollbarTrackStartInset", b, this.h0);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "jc", {
        get: function () {
            return this.hz;
        },
        set: function (a) {
            var b = this.hz;
            this.hz = a;
            if (b != this.hz) {
                this.t5("HorizontalViewScrollbarTrackEndInset", b, this.hz);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "gm", {
        get: function () {
            return this.fp;
        },
        set: function (a) {
            var b = this.fp;
            this.fp = a;
            if (b != this.fp) {
                this.t5(DomainChart.pt, b, this.fp);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "wr", {
        get: function () {
            return this.vw;
        },
        set: function (a) {
            var b = this.vw;
            this.vw = a;
            if (b != this.vw) {
                this.t5("VerticalViewScrollbarFill", b, this.vw);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "ws", {
        get: function () {
            return this.vx;
        },
        set: function (a) {
            var b = this.vx;
            this.vx = a;
            if (b != this.vx) {
                this.t5("VerticalViewScrollbarOutline", b, this.vx);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "j3", {
        get: function () {
            return this.is;
        },
        set: function (a) {
            var b = this.is;
            this.is = a;
            if (b != this.is) {
                this.t5("VerticalViewScrollbarStrokeThickness", b, this.is);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "j2", {
        get: function () {
            return this.ir;
        },
        set: function (a) {
            var b = this.ir;
            this.ir = a;
            if (b != this.ir) {
                this.t5("VerticalViewScrollbarMaxOpacity", b, this.ir);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "j0", {
        get: function () {
            return this.ip;
        },
        set: function (a) {
            var b = this.ip;
            this.ip = a;
            if (b != this.ip) {
                this.t5("VerticalViewScrollbarCornerRadius", b, this.ip);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "j6", {
        get: function () {
            return this.iv;
        },
        set: function (a) {
            var b = this.iv;
            this.iv = a;
            if (b != this.iv) {
                this.t5("VerticalViewScrollbarWidth", b, this.iv);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "j1", {
        get: function () {
            return this.iq;
        },
        set: function (a) {
            var b = this.iq;
            this.iq = a;
            if (b != this.iq) {
                this.t5("VerticalViewScrollbarInset", b, this.iq);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "j5", {
        get: function () {
            return this.iu;
        },
        set: function (a) {
            var b = this.iu;
            this.iu = a;
            if (b != this.iu) {
                this.t5("VerticalViewScrollbarTrackStartInset", b, this.iu);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "j4", {
        get: function () {
            return this.it;
        },
        set: function (a) {
            var b = this.it;
            this.it = a;
            if (b != this.it) {
                this.t5("VerticalViewScrollbarTrackEndInset", b, this.it);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "g7", {
        get: function () {
            return this.f5;
        },
        set: function (a) {
            var b = this.f5;
            this.f5 = a;
            if (b != this.f5) {
                this.t5(DomainChart.sf, b, this.f5);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "j8", {
        get: function () {
            return this.ix;
        },
        set: function (a) {
            var b = this.ix;
            this.ix = a;
            if (b != this.ix) {
                this.t5("WindowRectMinWidth", b, this.ix);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "j7", {
        get: function () {
            return this.iw;
        },
        set: function (a) {
            var b = this.iw;
            this.iw = a;
            if (b != this.iw) {
                this.t5("WindowRectMinHeight", b, this.iw);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "ka", {
        get: function () {
            return this.iz;
        },
        set: function (a) {
            var b = this.iz;
            this.iz = a;
            if (b != this.iz) {
                this.t5("WindowSizeMinWidth", b, this.iz);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChart.prototype, "j9", {
        get: function () {
            return this.iy;
        },
        set: function (a) {
            var b = this.iy;
            this.iy = a;
            if (b != this.iy) {
                this.t5("WindowSizeMinHeight", b, this.iy);
            }
        },
        enumerable: false,
        configurable: true
    });
    DomainChart.$t = markType(DomainChart, 'DomainChart', Base.$, [INotifyPropertyChanged_$type, IDataLegendTarget_$type, IToolbarTarget_$type]);
    DomainChart.k9 = new Random(0);
    DomainChart.k8 = new Random(0);
    DomainChart.nx = "DataToolTipShouldUpdateWhenSeriesDataChanges";
    DomainChart.pt = "HorizontalViewScrollbarShouldAddAutoTrackInsets";
    DomainChart.sf = "VerticalViewScrollbarShouldAddAutoTrackInsets";
    return DomainChart;
}(Base));
export { DomainChart };
/**
 * @hidden
 */
export var IDomainChartToolbarProvider_$type = /*@__PURE__*/ new Type(null, 'IDomainChartToolbarProvider');
