// igniteui-react-core/esm5/lib/SeriesMatcherDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var SeriesMatcherDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(SeriesMatcherDescription, _super);
    function SeriesMatcherDescription() {
        var _this = _super.call(this) || this;
        _this.l = null;
        _this.m = null;
        _this.h = 0;
        _this.k = null;
        _this.j = null;
        return _this;
    }
    SeriesMatcherDescription.prototype.get_type = function () {
        return "SeriesMatcher";
    };
    Object.defineProperty(SeriesMatcherDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesMatcherDescription.prototype, "name", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("Name");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesMatcherDescription.prototype, "title", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("Title");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesMatcherDescription.prototype, "index", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Index");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesMatcherDescription.prototype, "memberPathType", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("MemberPathType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesMatcherDescription.prototype, "memberPath", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("MemberPath");
        },
        enumerable: false,
        configurable: true
    });
    SeriesMatcherDescription.$t = markType(SeriesMatcherDescription, 'SeriesMatcherDescription', Description.$);
    return SeriesMatcherDescription;
}(Description));
export { SeriesMatcherDescription };
