// igniteui-react-core/esm5/lib/WebDateTimeInputDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { WebMaskInputBaseDescription } from "./WebMaskInputBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var WebDateTimeInputDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebDateTimeInputDescription, _super);
    function WebDateTimeInputDescription() {
        var _this = _super.call(this) || this;
        _this.a0 = null;
        _this.as = new Date();
        _this.aq = new Date();
        _this.ao = new Date();
        _this.ar = new Date();
        _this.ap = new Date();
        _this.az = null;
        _this.ak = null;
        _this.am = false;
        _this.a1 = null;
        _this.ay = null;
        return _this;
    }
    WebDateTimeInputDescription.prototype.get_type = function () {
        return "WebDateTimeInput";
    };
    Object.defineProperty(WebDateTimeInputDescription.prototype, "inputFormat", {
        get: function () {
            return this.a0;
        },
        set: function (a) {
            this.a0 = a;
            this.g("InputFormat");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDateTimeInputDescription.prototype, "value", {
        get: function () {
            return this.as;
        },
        set: function (a) {
            this.as = a;
            this.g("Value");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDateTimeInputDescription.prototype, "min", {
        get: function () {
            return this.aq;
        },
        set: function (a) {
            this.aq = a;
            this.g("Min");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDateTimeInputDescription.prototype, "max", {
        get: function () {
            return this.ao;
        },
        set: function (a) {
            this.ao = a;
            this.g("Max");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDateTimeInputDescription.prototype, "minValue", {
        get: function () {
            return this.ar;
        },
        set: function (a) {
            this.ar = a;
            this.g("MinValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDateTimeInputDescription.prototype, "maxValue", {
        get: function () {
            return this.ap;
        },
        set: function (a) {
            this.ap = a;
            this.g("MaxValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDateTimeInputDescription.prototype, "displayFormat", {
        get: function () {
            return this.az;
        },
        set: function (a) {
            this.az = a;
            this.g("DisplayFormat");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDateTimeInputDescription.prototype, "spinDelta", {
        get: function () {
            return this.ak;
        },
        set: function (a) {
            this.ak = a;
            this.g("SpinDelta");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDateTimeInputDescription.prototype, "spinLoop", {
        get: function () {
            return this.am;
        },
        set: function (a) {
            this.am = a;
            this.g("SpinLoop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDateTimeInputDescription.prototype, "locale", {
        get: function () {
            return this.a1;
        },
        set: function (a) {
            this.a1 = a;
            this.g("Locale");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDateTimeInputDescription.prototype, "changeRef", {
        get: function () {
            return this.ay;
        },
        set: function (a) {
            this.ay = a;
            this.g("ChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    WebDateTimeInputDescription.$t = markType(WebDateTimeInputDescription, 'WebDateTimeInputDescription', WebMaskInputBaseDescription.$);
    return WebDateTimeInputDescription;
}(WebMaskInputBaseDescription));
export { WebDateTimeInputDescription };
