// igniteui-react-core/esm2015/lib/AbsoluteVolumeOscillatorIndicatorDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicatorDescription } from "./StrategyBasedIndicatorDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let AbsoluteVolumeOscillatorIndicatorDescription = /*@__PURE__*/ (() => {
    class AbsoluteVolumeOscillatorIndicatorDescription extends StrategyBasedIndicatorDescription {
        get_type() {
            return "AbsoluteVolumeOscillatorIndicator";
        }
        constructor() {
            super();
            this.f8 = 0;
            this.f7 = 0;
        }
        get shortPeriod() {
            return this.f8;
        }
        set shortPeriod(a) {
            this.f8 = a;
            this.g("ShortPeriod");
        }
        get longPeriod() {
            return this.f7;
        }
        set longPeriod(a) {
            this.f7 = a;
            this.g("LongPeriod");
        }
    }
    AbsoluteVolumeOscillatorIndicatorDescription.$t = /*@__PURE__*/ markType(AbsoluteVolumeOscillatorIndicatorDescription, 'AbsoluteVolumeOscillatorIndicatorDescription', StrategyBasedIndicatorDescription.$);
    return AbsoluteVolumeOscillatorIndicatorDescription;
})();
