// igniteui-react-core/esm2015/lib/ChartSelectionDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let ChartSelectionDescription = /*@__PURE__*/ (() => {
    class ChartSelectionDescription extends Description {
        get_type() {
            return "ChartSelection";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.m = null;
            this.h = null;
            this.j = null;
        }
        get item() {
            return this.m;
        }
        set item(a) {
            this.m = a;
            this.g("Item");
        }
        get series() {
            return this.h;
        }
        set series(a) {
            this.h = a;
            this.g("Series");
        }
        get matcher() {
            return this.j;
        }
        set matcher(a) {
            this.j = a;
            this.g("Matcher");
        }
    }
    ChartSelectionDescription.$t = /*@__PURE__*/ markType(ChartSelectionDescription, 'ChartSelectionDescription', Description.$);
    ChartSelectionDescription.__marshalByValue = true;
    ChartSelectionDescription.__marshalByValueAlias = "ChartSelection";
    return ChartSelectionDescription;
})();
