// igniteui-react-core/esm2015/lib/NativeUIButton.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NativeUIComponent } from "./NativeUIComponent";
import { NativeUIDisplayDensity_$type } from "./NativeUIDisplayDensity";
import { enumGetBox, EnumUtil, markType } from "./type";
import { NativeUIButtonDisplayType_$type } from "./NativeUIButtonDisplayType";
/**
 * @hidden
 */
export let NativeUIButton = /*@__PURE__*/ (() => {
    class NativeUIButton extends NativeUIComponent {
        constructor() {
            super(...arguments);
            this.bh = null;
        }
        get_e() {
            return 1;
        }
        get bl() {
            let a = this.ae(1);
            return a == null ? 1 : EnumUtil.getEnumValue(NativeUIDisplayDensity_$type, a);
        }
        set bl(a) {
            this.a0(1, enumGetBox(NativeUIDisplayDensity_$type, a));
        }
        get bj() {
            let a = this.ae(0);
            return a == null ? 0 : EnumUtil.getEnumValue(NativeUIButtonDisplayType_$type, a);
        }
        set bj(a) {
            this.a0(0, enumGetBox(NativeUIButtonDisplayType_$type, a));
        }
        get bt() {
            let a = this.ae(3);
            return a;
        }
        set bt(a) {
            this.a0(3, a);
        }
        get bk() {
            return this.ae(54);
        }
        set bk(a) {
            this.a0(54, a);
        }
        get bq() {
            return this.ae(4);
        }
        set bq(a) {
            this.a0(4, a);
        }
        get bo() {
            return this.ae(5);
        }
        set bo(a) {
            this.a0(5, a);
        }
        get bn() {
            return this.ae(6);
        }
        set bn(a) {
            this.a0(6, a);
        }
        get bs() {
            return this.ae(32);
        }
        set bs(a) {
            this.a0(32, a);
        }
        get bv() {
            return this.ae(24);
        }
        set bv(a) {
            this.a0(24, a);
        }
        get bw() {
            return this.ae(55);
        }
        set bw(a) {
            this.a0(55, a);
        }
        get br() {
            return this.ae(57);
        }
        set br(a) {
            this.a0(57, a);
        }
        get bp() {
            return this.ae(56);
        }
        set bp(a) {
            this.a0(56, a);
        }
        get bm() {
            return this.ae(58);
        }
        set bm(a) {
            this.a0(58, a);
        }
        get bu() {
            return this.ae(17);
        }
        set bu(a) {
            this.a0(17, a);
        }
        get bg() {
            return this.ae(11);
        }
        set bg(a) {
            this.a0(11, a);
        }
        get bi() {
            return this.bh;
        }
        set bi(a) {
            let b = this.bh;
            this.bh = a;
            if (this.bh != b) {
                if (b != null) {
                    this.ay(1, b);
                }
                if (this.bh != null) {
                    this.an(1, this.bh);
                }
            }
        }
    }
    NativeUIButton.$t = /*@__PURE__*/ markType(NativeUIButton, 'NativeUIButton', NativeUIComponent.$);
    return NativeUIButton;
})();
