// igniteui-react-core/esm2015/lib/SeriesMatcherDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let SeriesMatcherDescription = /*@__PURE__*/ (() => {
    class SeriesMatcherDescription extends Description {
        get_type() {
            return "SeriesMatcher";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.l = null;
            this.m = null;
            this.h = 0;
            this.k = null;
            this.j = null;
        }
        get name() {
            return this.l;
        }
        set name(a) {
            this.l = a;
            this.g("Name");
        }
        get title() {
            return this.m;
        }
        set title(a) {
            this.m = a;
            this.g("Title");
        }
        get index() {
            return this.h;
        }
        set index(a) {
            this.h = a;
            this.g("Index");
        }
        get memberPathType() {
            return this.k;
        }
        set memberPathType(a) {
            this.k = a;
            this.g("MemberPathType");
        }
        get memberPath() {
            return this.j;
        }
        set memberPath(a) {
            this.j = a;
            this.g("MemberPath");
        }
    }
    SeriesMatcherDescription.$t = /*@__PURE__*/ markType(SeriesMatcherDescription, 'SeriesMatcherDescription', Description.$);
    return SeriesMatcherDescription;
})();
