// igniteui-react-core/esm2015/lib/ToolActionSeparatorDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ToolActionDescription } from "./ToolActionDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ToolActionSeparatorDescription = /*@__PURE__*/ (() => {
    class ToolActionSeparatorDescription extends ToolActionDescription {
        get_type() {
            return "ToolActionSeparator";
        }
        constructor() {
            super();
            this.c0 = 0;
            this.cy = false;
        }
        get size() {
            return this.c0;
        }
        set size(a) {
            this.c0 = a;
            this.g("Size");
        }
        get isGroupHeaderSeparator() {
            return this.cy;
        }
        set isGroupHeaderSeparator(a) {
            this.cy = a;
            this.g("IsGroupHeaderSeparator");
        }
    }
    ToolActionSeparatorDescription.$t = /*@__PURE__*/ markType(ToolActionSeparatorDescription, 'ToolActionSeparatorDescription', ToolActionDescription.$);
    return ToolActionSeparatorDescription;
})();
