// igniteui-react-core/esm2015/lib/ToolActionButtonDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ToolActionDescription } from "./ToolActionDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ToolActionButtonDescription = /*@__PURE__*/ (() => {
    class ToolActionButtonDescription extends ToolActionDescription {
        get_type() {
            return "ToolActionButton";
        }
        constructor() {
            super();
            this.c0 = null;
            this.cy = 0;
        }
        get displayType() {
            return this.c0;
        }
        set displayType(a) {
            this.c0 = a;
            this.g("DisplayType");
        }
        get cornerRadius() {
            return this.cy;
        }
        set cornerRadius(a) {
            this.cy = a;
            this.g("CornerRadius");
        }
    }
    ToolActionButtonDescription.$t = /*@__PURE__*/ markType(ToolActionButtonDescription, 'ToolActionButtonDescription', ToolActionDescription.$);
    return ToolActionButtonDescription;
})();
