// igniteui-react-core/esm2015/lib/WebColumnStateDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebColumnStateDescription = /*@__PURE__*/ (() => {
    class WebColumnStateDescription extends Description {
        get_type() {
            return "WebColumnState";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.u = false;
            this.x = false;
            this.o = false;
            this.m = false;
            this.y = false;
            this.p = false;
            this.a0 = null;
            this.a1 = null;
            this.a2 = null;
            this.q = false;
            this.t = false;
            this.s = false;
            this.ax = null;
            this.r = false;
            this.ay = null;
            this.at = null;
            this.az = null;
            this.v = false;
            this.w = false;
            this.j = false;
            this.as = null;
            this.k = false;
            this.l = false;
            this.i = false;
            this.n = false;
            this.z = false;
        }
        get pinned() {
            return this.u;
        }
        set pinned(a) {
            this.u = a;
            this.g("Pinned");
        }
        get sortable() {
            return this.x;
        }
        set sortable(a) {
            this.x = a;
            this.g("Sortable");
        }
        get filterable() {
            return this.o;
        }
        set filterable(a) {
            this.o = a;
            this.g("Filterable");
        }
        get editable() {
            return this.m;
        }
        set editable(a) {
            this.m = a;
            this.g("Editable");
        }
        get sortingIgnoreCase() {
            return this.y;
        }
        set sortingIgnoreCase(a) {
            this.y = a;
            this.g("SortingIgnoreCase");
        }
        get filteringIgnoreCase() {
            return this.p;
        }
        set filteringIgnoreCase(a) {
            this.p = a;
            this.g("FilteringIgnoreCase");
        }
        get headerClasses() {
            return this.a0;
        }
        set headerClasses(a) {
            this.a0 = a;
            this.g("HeaderClasses");
        }
        get headerGroupClasses() {
            return this.a1;
        }
        set headerGroupClasses(a) {
            this.a1 = a;
            this.g("HeaderGroupClasses");
        }
        get maxWidth() {
            return this.a2;
        }
        set maxWidth(a) {
            this.a2 = a;
            this.g("MaxWidth");
        }
        get groupable() {
            return this.q;
        }
        set groupable(a) {
            this.q = a;
            this.g("Groupable");
        }
        get movable() {
            return this.t;
        }
        set movable(a) {
            this.t = a;
            this.g("Movable");
        }
        get hidden() {
            return this.s;
        }
        set hidden(a) {
            this.s = a;
            this.g("Hidden");
        }
        get dataType() {
            return this.ax;
        }
        set dataType(a) {
            this.ax = a;
            this.g("DataType");
        }
        get hasSummary() {
            return this.r;
        }
        set hasSummary(a) {
            this.r = a;
            this.g("HasSummary");
        }
        get field() {
            return this.ay;
        }
        set field(a) {
            this.ay = a;
            this.g("Field");
        }
        get width() {
            return this.at;
        }
        set width(a) {
            this.at = a;
            this.g("Width");
        }
        get header() {
            return this.az;
        }
        set header(a) {
            this.az = a;
            this.g("Header");
        }
        get resizable() {
            return this.v;
        }
        set resizable(a) {
            this.v = a;
            this.g("Resizable");
        }
        get searchable() {
            return this.w;
        }
        set searchable(a) {
            this.w = a;
            this.g("Searchable");
        }
        get columnGroup() {
            return this.j;
        }
        set columnGroup(a) {
            this.j = a;
            this.g("ColumnGroup");
        }
        get parent() {
            return this.as;
        }
        set parent(a) {
            this.as = a;
            this.g("Parent");
        }
        get disableHiding() {
            return this.k;
        }
        set disableHiding(a) {
            this.k = a;
            this.g("DisableHiding");
        }
        get disablePinning() {
            return this.l;
        }
        set disablePinning(a) {
            this.l = a;
            this.g("DisablePinning");
        }
        get collapsible() {
            return this.i;
        }
        set collapsible(a) {
            this.i = a;
            this.g("Collapsible");
        }
        get expanded() {
            return this.n;
        }
        set expanded(a) {
            this.n = a;
            this.g("Expanded");
        }
        get visibleWhenCollapsed() {
            return this.z;
        }
        set visibleWhenCollapsed(a) {
            this.z = a;
            this.g("VisibleWhenCollapsed");
        }
    }
    WebColumnStateDescription.$t = /*@__PURE__*/ markType(WebColumnStateDescription, 'WebColumnStateDescription', Description.$);
    WebColumnStateDescription.__marshalByValue = true;
    WebColumnStateDescription.__marshalByValueAlias = "ColumnState";
    return WebColumnStateDescription;
})();
