// igniteui-react-charts/esm2015/lib/AccumulationDistributionIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { AccumulationDistributionIndicatorStrategy } from "./AccumulationDistributionIndicatorStrategy";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let AccumulationDistributionIndicator = /*@__PURE__*/ (() => {
    class AccumulationDistributionIndicator extends StrategyBasedIndicator {
        bb() {
            return new AccumulationDistributionIndicator();
        }
        get_aaa() {
            return new AccumulationDistributionIndicatorStrategy();
        }
        get aaa() {
            return this.get_aaa();
        }
        get_aan() {
            return AccumulationDistributionIndicator.$;
        }
        get aan() {
            return this.get_aan();
        }
    }
    AccumulationDistributionIndicator.$t = /*@__PURE__*/ markType(AccumulationDistributionIndicator, 'AccumulationDistributionIndicator', StrategyBasedIndicator.$);
    return AccumulationDistributionIndicator;
})();
