// igniteui-react-charts/esm2015/lib/EaseOfMovementIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { EaseOfMovementIndicatorStrategy } from "./EaseOfMovementIndicatorStrategy";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let EaseOfMovementIndicator = /*@__PURE__*/ (() => {
    class EaseOfMovementIndicator extends StrategyBasedIndicator {
        bb() {
            return new EaseOfMovementIndicator();
        }
        get_aaa() {
            return new EaseOfMovementIndicatorStrategy();
        }
        get aaa() {
            return this.get_aaa();
        }
        get_aan() {
            return EaseOfMovementIndicator.$;
        }
        get aan() {
            return this.get_aan();
        }
    }
    EaseOfMovementIndicator.$t = /*@__PURE__*/ markType(EaseOfMovementIndicator, 'EaseOfMovementIndicator', StrategyBasedIndicator.$);
    return EaseOfMovementIndicator;
})();
