// igniteui-react-charts/esm2015/lib/igr-value-layer.js
import { ValueLayerValueMode_$type } from "./ValueLayerValueMode";
import { IgrSeries } from "./igr-series";
import { IgrAxis } from "./igr-axis";
import { IgrAnnotationLayer } from "./igr-annotation-layer";
import { ValueLayer } from "./ValueLayer";
import { ensureBool, ensureEnum, brushToString, stringToBrush } from "igniteui-react-core";
/**
 * Represents an annotation layer that displays crosshair lines that cross through the closest value of the target series under the cursor.
*/
export class IgrValueLayer extends IgrAnnotationLayer {
    createImplementation() {
        return new ValueLayer();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor(props) {
        super(props);
    }
    /**
     * Gets whether the series is an crosshair annotation layer.
    */
    get isAnnotationValueLayer() {
        return this.i.ee;
    }
    /**
     * Gets or sets the value mode for the overlay.
    */
    get valueMode() {
        return this.i.yh;
    }
    set valueMode(v) {
        this.i.yh = ensureEnum(ValueLayerValueMode_$type, v);
    }
    /**
     * Gets or sets the color to use for the horizontal line. Leave null for an automatic value.
    */
    get horizontalLineStroke() {
        return brushToString(this.i.aa5);
    }
    set horizontalLineStroke(v) {
        this.i.aa5 = stringToBrush(v);
    }
    /**
     * Gets or sets the color to use for vertical line. Leave null for an automatic value.
    */
    get verticalLineStroke() {
        return brushToString(this.i.aa6);
    }
    set verticalLineStroke(v) {
        this.i.aa6 = stringToBrush(v);
    }
    /**
     * Gets or sets the name of the series to target this annotation to. If null, this annotation targets all series simultaneously.
    */
    get targetSeriesName() {
        return this.i.za;
    }
    set targetSeriesName(v) {
        this.i.za = v;
    }
    /**
     * Gets or sets the series to target this annotation to. If null, this annotation targets all series simultaneously.
    */
    get targetSeries() {
        const r = this.i.yf;
        if (r == null) {
            return null;
        }
        if (!r.externalObject) {
            let e = IgrSeries._createFromInternal(r);
            if (e) {
                e._implementation = r;
            }
            r.externalObject = e;
        }
        return r.externalObject;
    }
    set targetSeries(v) {
        if (v != null && this._stylingContainer && v._styling)
            v._styling(this._stylingContainer, this, this);
        v == null ? this.i.yf = null : this.i.yf = v.i;
    }
    /**
     * Gets or sets the name of the axis to target this annotation to. If null, this annotation targets all axis simultaneously.
    */
    get targetAxisName() {
        return this.i.y7;
    }
    set targetAxisName(v) {
        this.i.y7 = v;
    }
    /**
     * Gets or sets the axis to target this annotation to. If null, this annotation targets all value axis simultaneously.
    */
    get targetAxis() {
        const r = this.i.yc;
        if (r == null) {
            return null;
        }
        if (!r.externalObject) {
            let e = IgrAxis._createFromInternal(r);
            if (e) {
                e._implementation = r;
            }
            r.externalObject = e;
        }
        return r.externalObject;
    }
    set targetAxis(v) {
        if (v != null && this._stylingContainer && v._styling)
            v._styling(this._stylingContainer, this, this);
        v == null ? this.i.yc = null : this.i.yc = v.i;
    }
    /**
     * Gets or sets whether to use value interpolation when drawing a line through the best value for the cursor position.
    */
    get useInterpolation() {
        return this.i.ym;
    }
    set useInterpolation(v) {
        this.i.ym = ensureBool(v);
    }
    /**
     * Gets or sets whether to draw annotations over the axes where the crosshair meets with them.
    */
    get isAxisAnnotationEnabled() {
        return this.i.yj;
    }
    set isAxisAnnotationEnabled(v) {
        this.i.yj = ensureBool(v);
    }
    /**
     * Sets or gets a function which takes an object that produces a formatted label for the x axis annotation.
    */
    get xAxisAnnotationFormatLabel() {
        return this.i.yd;
    }
    set xAxisAnnotationFormatLabel(v) {
        this.i.yd = v;
    }
    /**
     * Sets or gets a function which takes an object that produces a formatted label for the y axis annotation.
    */
    get yAxisAnnotationFormatLabel() {
        return this.i.ye;
    }
    set yAxisAnnotationFormatLabel(v) {
        this.i.ye = v;
    }
    /**
     * Gets or sets the color to use for the x axis annotation text. Leave unset for an automatic value.
    */
    get xAxisAnnotationTextColor() {
        return brushToString(this.i.aa9);
    }
    set xAxisAnnotationTextColor(v) {
        this.i.aa9 = stringToBrush(v);
    }
    /**
     * Gets or sets the color to use for the x axis annotation backing. Leave unset for an automatic value.
    */
    get xAxisAnnotationBackground() {
        return brushToString(this.i.aa7);
    }
    set xAxisAnnotationBackground(v) {
        this.i.aa7 = stringToBrush(v);
    }
    /**
     * Gets or sets the corner radius to use for the x axis annotation backing. Leave unset for an automatic value.
    */
    get xAxisAnnotationBackgroundCornerRadius() {
        return this.i.yp;
    }
    set xAxisAnnotationBackgroundCornerRadius(v) {
        this.i.yp = +v;
    }
    /**
     * Gets or sets the precision to use displaying values for interpolated crosshair positions.
    */
    get xAxisAnnotationInterpolatedValuePrecision() {
        return this.i.y1;
    }
    set xAxisAnnotationInterpolatedValuePrecision(v) {
        this.i.y1 = +v;
    }
    /**
     * Gets or sets the color to use for the x axis annotation outline. Leave unset for an automatic value.
    */
    get xAxisAnnotationOutline() {
        return brushToString(this.i.aa8);
    }
    set xAxisAnnotationOutline(v) {
        this.i.aa8 = stringToBrush(v);
    }
    /**
     * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
    */
    get xAxisAnnotationPaddingLeft() {
        return this.i.yr;
    }
    set xAxisAnnotationPaddingLeft(v) {
        this.i.yr = +v;
    }
    /**
     * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
    */
    get xAxisAnnotationPaddingTop() {
        return this.i.yt;
    }
    set xAxisAnnotationPaddingTop(v) {
        this.i.yt = +v;
    }
    /**
     * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
    */
    get xAxisAnnotationPaddingRight() {
        return this.i.ys;
    }
    set xAxisAnnotationPaddingRight(v) {
        this.i.ys = +v;
    }
    /**
     * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
    */
    get xAxisAnnotationPaddingBottom() {
        return this.i.yq;
    }
    set xAxisAnnotationPaddingBottom(v) {
        this.i.yq = +v;
    }
    /**
     * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
    */
    get yAxisAnnotationPaddingLeft() {
        return this.i.yx;
    }
    set yAxisAnnotationPaddingLeft(v) {
        this.i.yx = +v;
    }
    /**
     * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
    */
    get yAxisAnnotationPaddingTop() {
        return this.i.yz;
    }
    set yAxisAnnotationPaddingTop(v) {
        this.i.yz = +v;
    }
    /**
     * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
    */
    get yAxisAnnotationPaddingRight() {
        return this.i.yy;
    }
    set yAxisAnnotationPaddingRight(v) {
        this.i.yy = +v;
    }
    /**
     * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
    */
    get yAxisAnnotationPaddingBottom() {
        return this.i.yw;
    }
    set yAxisAnnotationPaddingBottom(v) {
        this.i.yw = +v;
    }
    /**
     * Gets or sets the stroke thickness for the x axis annotation backing. Leave unset for an automatic value.
    */
    get xAxisAnnotationStrokeThickness() {
        return this.i.yu;
    }
    set xAxisAnnotationStrokeThickness(v) {
        this.i.yu = +v;
    }
    /**
     * Gets or sets the color to use for the y axis annotation text. Leave unset for an automatic value.
    */
    get yAxisAnnotationTextColor() {
        return brushToString(this.i.abc);
    }
    set yAxisAnnotationTextColor(v) {
        this.i.abc = stringToBrush(v);
    }
    /**
     * Gets or sets the color to use for the y axis annotation backing. Leave unset for an automatic value.
    */
    get yAxisAnnotationBackground() {
        return brushToString(this.i.aba);
    }
    set yAxisAnnotationBackground(v) {
        this.i.aba = stringToBrush(v);
    }
    /**
     * Gets or sets the corner radius to use for the y axis annotation backing. Leave unset for an automatic value.
    */
    get yAxisAnnotationBackgroundCornerRadius() {
        return this.i.yv;
    }
    set yAxisAnnotationBackgroundCornerRadius(v) {
        this.i.yv = +v;
    }
    /**
     * Gets or sets the precision to use displaying values for interpolated crosshair positions.
    */
    get yAxisAnnotationInterpolatedValuePrecision() {
        return this.i.y2;
    }
    set yAxisAnnotationInterpolatedValuePrecision(v) {
        this.i.y2 = +v;
    }
    /**
     * Gets or sets the color to use for the y axis annotation outline. Leave unset for an automatic value.
    */
    get yAxisAnnotationOutline() {
        return brushToString(this.i.abb);
    }
    set yAxisAnnotationOutline(v) {
        this.i.abb = stringToBrush(v);
    }
    /**
     * Gets or sets the stroke thickness for the y axis annotation backing. Leave unset for an automatic value.
    */
    get yAxisAnnotationStrokeThickness() {
        return this.i.y0;
    }
    set yAxisAnnotationStrokeThickness(v) {
        this.i.y0 = +v;
    }
    /**
     * Gets or sets whether to skip unknown values when searching for series values.
    */
    get skipUnknownValues() {
        return this.i.yl;
    }
    set skipUnknownValues(v) {
        this.i.yl = ensureBool(v);
    }
    findByName(name) {
        var baseResult = super.findByName(name);
        if (baseResult) {
            return baseResult;
        }
        if (this.targetSeries && this.targetSeries.name && this.targetSeries.name == name) {
            return this.targetSeries;
        }
        if (this.targetAxis && this.targetAxis.name && this.targetAxis.name == name) {
            return this.targetAxis;
        }
        return null;
    }
    _styling(container, component, parent) {
        super._styling(container, component, parent);
        this._inStyling = true;
        if (this.targetSeries && this.targetSeries._styling) {
            this.targetSeries._styling(container, component, this);
        }
        if (this.targetAxis && this.targetAxis._styling) {
            this.targetAxis._styling(container, component, this);
        }
        this._inStyling = false;
    }
}
