// igniteui-react-charts/esm2015/lib/ItemwiseStrategyBasedIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { ItemwiseStrategyCalculationStrategy } from "./ItemwiseStrategyCalculationStrategy";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let ItemwiseStrategyBasedIndicator = /*@__PURE__*/ (() => {
    class ItemwiseStrategyBasedIndicator extends StrategyBasedIndicator {
        get_aaa() {
            return new ItemwiseStrategyCalculationStrategy();
        }
        get aaa() {
            return this.get_aaa();
        }
        aax() {
            return true;
        }
        constructor() {
            super();
        }
        aao() {
            super.aao();
            this.aav = this.aaw;
            this.z9.c = this.aav;
        }
    }
    ItemwiseStrategyBasedIndicator.$t = /*@__PURE__*/ markType(ItemwiseStrategyBasedIndicator, 'ItemwiseStrategyBasedIndicator', StrategyBasedIndicator.$);
    return ItemwiseStrategyBasedIndicator;
})();
