// igniteui-react-charts/esm2015/lib/OnBalanceVolumeIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { OnBalanceVolumeIndicatorStrategy } from "./OnBalanceVolumeIndicatorStrategy";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let OnBalanceVolumeIndicator = /*@__PURE__*/ (() => {
    class OnBalanceVolumeIndicator extends StrategyBasedIndicator {
        bb() {
            return new OnBalanceVolumeIndicator();
        }
        get_aaa() {
            return new OnBalanceVolumeIndicatorStrategy();
        }
        get aaa() {
            return this.get_aaa();
        }
        get_aan() {
            return OnBalanceVolumeIndicator.$;
        }
        get aan() {
            return this.get_aan();
        }
    }
    OnBalanceVolumeIndicator.$t = /*@__PURE__*/ markType(OnBalanceVolumeIndicator, 'OnBalanceVolumeIndicator', StrategyBasedIndicator.$);
    return OnBalanceVolumeIndicator;
})();
