// igniteui-react-charts/esm2015/lib/PositiveVolumeIndexIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { PositiveVolumeIndexIndicatorStrategy } from "./PositiveVolumeIndexIndicatorStrategy";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let PositiveVolumeIndexIndicator = /*@__PURE__*/ (() => {
    class PositiveVolumeIndexIndicator extends StrategyBasedIndicator {
        bb() {
            return new PositiveVolumeIndexIndicator();
        }
        get_aaa() {
            return new PositiveVolumeIndexIndicatorStrategy();
        }
        get aaa() {
            return this.get_aaa();
        }
        get_aan() {
            return PositiveVolumeIndexIndicator.$;
        }
        get aan() {
            return this.get_aan();
        }
    }
    PositiveVolumeIndexIndicator.$t = /*@__PURE__*/ markType(PositiveVolumeIndexIndicator, 'PositiveVolumeIndexIndicator', StrategyBasedIndicator.$);
    return PositiveVolumeIndexIndicator;
})();
